{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "notes"
    }
   },
   "source": [
    "# Use NCBI Datasets to Retrieve Information on Genes\n",
    "\n",
    "The objective of this notebook is to use the `ncbi.datasets` python library to download and extract gene data.\n",
    "\n",
    "In this example, we will get information about the Gonadotropin Releasing Hormone Receptor (GNRHR) gene family, which plays a key role in sexual development and function across vertebrates.\n",
    "\n",
    "This notebook will demonstrate the following features that are available through NCBI Datasets for retrieving gene data:  \n",
    "1. Getting gene metadata as a JSON-formatted gene descriptor, by specifying either an NCBI Gene ID or gene symbol  \n",
    "2. Downloading gene datasets, which include gene, transcript, and protein sequences, and a data report (gene metadata in a more human-readable yaml format)\n",
    "\n",
    "Gonadotropin Releasing Hormone (GnRH) binds and activates the GnRH receptor (GnRHR) to stimulate secretion of follicle stimulating hormone (FSH) and luteinizing hormones (LH). These hormones in turn regulate repoductive development and function. \n",
    "\n",
    "While the role of GnRH in reproduction is conserved across vertebrates, GnRH receptor copy number is variable. \n",
    "In humans and some primates, there are two GnRHR genes, while in fish and amphibians, three GnRHR genes have been identified, with additional duplications observed in some fish. Additional receptors and GnRH ligands suggests that new functions could have been acquired by the additional gene copies[1].\n",
    "\n",
    "[1] Moncaut N, Somoza G, Power DM, CanÃ¡rio AV. Five gonadotrophin-releasing hormone receptors in a teleost fish: isolation, tissue distribution and phylogenetic relationships. J Mol Endocrinol. 2005 Jun;34(3):767-79. doi: 10.1677/jme.1.01757. PMID: 15956346.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "notes"
    }
   },
   "source": [
    "First we will load the libraries necessary to run this notebook. This includes:\n",
    "1. the ncbi.datasets library \n",
    "2. pandas, the python data analysis library and \n",
    "3. pprint, which allows \"pretty-printing\" of Python data structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "slideshow": {
     "slide_type": "skip"
    }
   },
   "outputs": [],
   "source": [
    "# load all libraries\n",
    "import pandas as pd\n",
    "import ncbi.datasets\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Get gene descriptors for three human GnRHR genes \n",
    "First we're going to get gene descriptors (metadata) for three human GnRHR genes, GNRHR, GNRHR2, and GNRHR2P1, by specifying the NCBI Gene IDs for these genes. \n",
    "\n",
    "Gene descriptors provide gene metadata in a machine-readable JSON format.\n",
    "\n",
    "Gene descriptors contain a lot of interesting metadata and it's easy to pull out just the fields that you're interested in. As an example, here we'll show gene symbols, chromosome number and the corresponding SwissProt accession for the three genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t114814 -> GNRHR2 (Chromosome(s): ['1'], SwissProt: ['Q96P88.4'])\n",
      "\t2798 -> GNRHR (Chromosome(s): ['4'], SwissProt: ['P30968.1'])\n",
      "\t404718 -> GNRHR2P1 (Chromosome(s): ['14'], SwissProt: None)\n"
     ]
    }
   ],
   "source": [
    "# start a datasets gene API instance\n",
    "api_client = ncbi.datasets.ApiClient()\n",
    "ds_gene_instance = ncbi.datasets.GeneApi(api_client)\n",
    "\n",
    "# Retrieve descriptors for three known gene-ids.\n",
    "genes_and_messages = ds_gene_instance.gene_metadata_by_id([2798, 114814, 404718])\n",
    "\n",
    "# Look up the symbols for each gene\n",
    "def report_on_gene_descriptors(genes_and_messages, leader='\\t', report_errors=True):\n",
    "    if report_errors:\n",
    "        for message in genes_and_messages.messages or []:\n",
    "            print(f'{leader}Error for: ({\",\".join(message.error.invalid_identifiers)})')\n",
    "            print(f'{leader}{leader}Reason: ({message.error.reason})')\n",
    "\n",
    "    if not genes_and_messages.genes:\n",
    "        print(f'{leader}Gene-ids not found in this organism')\n",
    "        return\n",
    "\n",
    "    for gene in map(lambda g: g.gene, genes_and_messages.genes):\n",
    "        print(f'{leader}{gene.gene_id} -> {gene.symbol} (Chromosome(s): {gene.chromosomes}, SwissProt: {gene.swiss_prot_accessions})')\n",
    "\n",
    "report_on_gene_descriptors(genes_and_messages)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get gene descriptors for GNRHR genes throughout primates\n",
    "Now we're going to get gene descriptors (metadata) for a large set of GnRHR genes in primates. To get these gene descriptors we can specify gene symbol and species-level Tax IDs.  \n",
    "In order to get these species-level Tax ID, we're going to use the Gene API to get the species-level TaxIDs under the taxonomic rank of primate.\n",
    "We're going to search for two symbols, GNRHR and GNRHR2. Interestingly, very few primates have an annotated gene ortholog of human GNRHR2, and this is reflected in the results. In fact, the only non-human primates for which we can find GNRHR2 is in Rhesus monkey (Macaca mulatta) and white-tufted-ear marmoset.\n",
    "For those primates without an apparent GNRHR2, it's also possible that we might miss the gene because of inconsistent (or missing) nomenclature.\n",
    "In a future iteration of Datasets, we plan to give users access to orthology information that would allow retrieval of a more comprehensive set of gene homologs across organisms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetch for eastern hoolock gibbon (Hoolock leuconedys / 593543)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for hoolock gibbon (Hoolock hoolock / 61851)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Hoolock leuconedys x Hoolock tianxing / 1934191)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Skywalker hoolock gibbon (Hoolock tianxing / 1934190)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for agile gibbon (Hylobates agilis / 9579)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for common gibbon (Hylobates lar / 9580)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for pileated gibbon (Hylobates pileatus / 9589)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for silvery gibbon (Hylobates moloch / 81572)\n",
      "\t116468321 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for northern white-cheeked gibbon (Nomascus leucogenys / 61853)\n",
      "\t100589627 -> GNRHR (Chromosome(s): ['9'], SwissProt: None)\n",
      "Fetch for Red-cheeked gibbon (Nomascus gabriellae / 61852)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for southern white-cheeked gibbon (Nomascus siki / 9586)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for siamang (Symphalangus syndactylus / 9590)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for chimpanzee (Pan troglodytes / 9598)\n",
      "\t471226 -> GNRHR (Chromosome(s): ['4'], SwissProt: None)\n",
      "Fetch for pygmy chimpanzee (Pan paniscus / 9597)\n",
      "\t100967998 -> GNRHR (Chromosome(s): ['4'], SwissProt: None)\n",
      "Fetch for eastern gorilla (Gorilla beringei / 499232)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for western gorilla (Gorilla gorilla / 9593)\n",
      "\t101126857 -> GNRHR (Chromosome(s): ['4'], SwissProt: None)\n",
      "Fetch for Heidelberg man (Homo heidelbergensis / 1425170)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for human (Homo sapiens / 9606)\n",
      "\t114814 -> GNRHR2 (Chromosome(s): ['1'], SwissProt: ['Q96P88.4'])\n",
      "\t2798 -> GNRHR (Chromosome(s): ['4'], SwissProt: ['P30968.1'])\n",
      "Fetch for Bornean orangutan (Pongo pygmaeus / 9600)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Sumatran orangutan (Pongo abelii / 9601)\n",
      "\t100437372 -> GNRHR (Chromosome(s): ['4'], SwissProt: None)\n",
      "Fetch for Allen's swamp monkey (Allenopithecus nigroviridis / 54135)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for L'Hoest's monkey (Allochrocebus lhoesti / 100224)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Chacma baboon (Papio ursinus / 36229)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Guinea baboon (Papio papio / 100937)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for hamadryas baboon (Papio hamadryas / 9557)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Kinda baboon (Papio kindae / 208091)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for olive baboon (Papio anubis / 9555)\n",
      "\t100998025 -> GNRHR (Chromosome(s): ['3'], SwissProt: None)\n",
      "Fetch for Yellow baboon (Papio cynocephalus / 9556)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for collared mangabey (Cercocebus torquatus / 9530)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Golden-bellied mangabey (Cercocebus chrysogaster / 75569)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for sooty mangabey (Cercocebus atys / 9531)\n",
      "\t105596089 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for blue monkey (Cercopithecus mitis / 36225)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Diana monkey (Cercopithecus diana / 36224)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Sykes' monkey (Cercopithecus albogularis / 867370)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Bale Mountains vervet (Chlorocebus djamdjamensis / 1284215)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Chlorocebus aethiops x Chlorocebus pygerythrus / 1903316)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for green monkey (Chlorocebus sabaeus / 60711)\n",
      "\t103235600 -> GNRHR (Chromosome(s): ['7'], SwissProt: None)\n",
      "Fetch for grivet (Chlorocebus aethiops / 9534)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for malbrouck (Chlorocebus cynosuros / 460675)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for tantalus monkey (Chlorocebus tantalus / 60712)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for vervet (Chlorocebus pygerythrus / 60710)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for red guenon (Erythrocebus patas / 9538)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for drill (Mandrillus leucophaeus / 9568)\n",
      "\t105528593 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for mandrill (Mandrillus sphinx / 9561)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Black crested mangabey (Lophocebus aterrimus / 75566)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Assam macaque (Macaca assamensis / 9551)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Barbary ape (Macaca sylvanus / 9546)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for bonnet macaque (Macaca radiata / 9548)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Celebes crested macaque (Macaca nigra / 54600)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for crab-eating macaque (Macaca fascicularis / 9541)\n",
      "\t102133556 -> GNRHR (Chromosome(s): ['5'], SwissProt: None)\n",
      "Fetch for Japanese macaque (Macaca fuscata / 9542)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for liontail macaque (Macaca silenus / 54601)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Macaca leucogenys / 1885588)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Northern pig-tailed macaque (Macaca leonina / 90387)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Pere David's macaque (Macaca thibetana / 54602)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for pig-tailed macaque (Macaca nemestrina / 9545)\n",
      "\t105463605 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for Rhesus monkey (Macaca mulatta / 9544)\n",
      "\t574163 -> GNRHR2 (Chromosome(s): ['1'], SwissProt: ['Q95JG1.1'])\n",
      "\t711577 -> GNRHR (Chromosome(s): ['5'], SwissProt: ['Q8SPZ1.1'])\n",
      "Fetch for stump-tailed macaque (Macaca arctoides / 9540)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Taiwan macaque (Macaca cyclopis / 78449)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Tonkean macaque (Macaca tonkeana / 40843)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for toque macaque (Macaca sinica / 9552)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for gelada (Theropithecus gelada / 9565)\n",
      "\t112625046 -> GNRHR (Chromosome(s): ['5'], SwissProt: None)\n",
      "Fetch for Angolan colobus (Colobus angolensis / 54131)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for mantled guereza (Colobus guereza / 33548)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Dusky leaf-monkey (Trachypithecus obscurus / 54181)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Francois's langur (Trachypithecus francoisi / 54180)\n",
      "\t117085552 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for Hatinh langur (Trachypithecus hatinhensis / 867383)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Indochinese lutung (Trachypithecus germaini / 271260)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Nilgiri langur (Trachypithecus johnii / 66063)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for purple-faced leaf monkey (Trachypithecus vetulus / 54137)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Shortridge's langur (Trachypithecus shortridgei / 1042121)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Silvery lutung (Trachypithecus cristatus / 122765)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Trachypithecus pileatus / 164651)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for White-headed langur (Trachypithecus poliocephalus / 465719)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for proboscis monkey (Nasalis larvatus / 43780)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Ugandan red Colobus (Piliocolobus tephrosceles / 591936)\n",
      "\t111555805 -> GNRHR (Chromosome(s): ['3'], SwissProt: None)\n",
      "Fetch for western red colobus (Piliocolobus badius / 164648)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for mitred leaf monkey (Presbytis melalophos / 78451)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for olive colobus (Procolobus verus / 373033)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Black-shanked douc langur (Pygathrix nigripes / 310352)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Pygathrix cinerea 1 RL-2012 / 1194332)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Pygathrix cinerea 2 RL-2012 / 1194333)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Red shanked douc langur (Pygathrix nemaeus / 54133)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for black snub-nosed monkey (Rhinopithecus bieti / 61621)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t108530291 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for Burmese snub-nosed monkey (Rhinopithecus strykeri / 1194336)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for golden snub-nosed monkey (Rhinopithecus roxellana / 61622)\n",
      "\t104664117 -> GNRHR (Chromosome(s): ['2'], SwissProt: None)\n",
      "Fetch for Gray snub-nosed monkey (Rhinopithecus brelichi / 224329)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Rhinopithecus bieti 1 RL-2012 / 1194334)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Rhinopithecus bieti 2 RL-2012 / 1194335)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Tonkin snub-nosed monkey (Rhinopithecus avunculus / 66062)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Hanuman langur (Semnopithecus entellus / 88029)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Simakobou (Simias concolor / 170207)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Azara's night monkey (Aotus azarai / 30591)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for gray-bellied night monkey (Aotus lemurinus / 43147)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Ma's night monkey (Aotus nancymaae / 37293)\n",
      "\t105731588 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for black howler monkey (Alouatta caraya / 9502)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for howler monkey (Alouatta seniculus / 9503)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for white-bellied spider monkey (Ateles belzebuth / 9507)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for brown woolly monkey (Lagothrix lagotricha / 9519)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for white-faced sapajou (Cebus capucinus / 9516)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for white-fronted capuchin (Cebus albifrons / 9514)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Tufted capuchin (Sapajus apella / 9515)\n",
      "\t116537243 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for yellow-breasted capuchin (Sapajus xanthosternos / 174599)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Goeldi's marmoset (Callimico goeldii / 9495)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for black-pencilled marmoset (Callithrix penicillata / 57378)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Geoffroy's marmoset (Callithrix geoffroyi / 52231)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for white-tufted-ear marmoset (Callithrix jacchus / 9483)\n",
      "\t100385305 -> GNRHR (Chromosome(s): ['3'], SwissProt: None)\n",
      "\t100399755 -> LOC100399755 (Chromosome(s): ['18'], SwissProt: ['Q95MG6.1'])\n",
      "Fetch for Wied's marmoset (Callithrix kuhlii / 867363)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for pygmy marmoset (Callithrix pygmaea / 9493)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for golden lion tamarin (Leontopithecus rosalia / 30588)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for golden-rumped lion tamarin (Leontopithecus chrysopygus / 58710)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for cotton-top tamarin (Saguinus oedipus / 9490)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Bolivian squirrel monkey (Saimiri boliviensis / 27679)\n",
      "\t101052255 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for Central American squirrel monkey (Saimiri oerstedii / 70928)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for common squirrel monkey (Saimiri sciureus / 9521)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for black titi (Cheracebus lugens / 210166)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Bolivian titi (Plecturocebus donacophilus / 230833)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for coppery titi (Plecturocebus cupreus / 202457)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Bald Uacari (Cacajao calvus / 30596)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Chiropotes albinasus / 198627)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Chiropotes israelita / 280163)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Philippine tarsier (Carlito syrichta / 1868482)\n",
      "\t103273704 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for Horsfield's tarsier (Cephalopachus bancanus / 9477)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Dian's tarsier (Tarsius dentatus / 449501)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Lariang tarsier (Tarsius lariang / 630277)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Tarsius wallacei / 981131)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for aye-aye (Daubentonia madagascariensis / 31869)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for hairy-eared dwarf lemur (Allocebus trichotis / 122248)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Crossley's dwarf lemur (Cheirogaleus crossleyi / 291259)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Greater dwarf lemur (Cheirogaleus major / 47177)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Lesser dwarf lemur (Cheirogaleus medius / 9460)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Sibree's dwarf lemur (Cheirogaleus sibreei / 752538)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Anosy mouse lemur (Microcebus tanosi / 1310123)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Arnhold's mouse lemur (Microcebus arnholdi / 864580)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Berthe's mouse lemur (Microcebus berthae / 143352)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for brown mouse lemur (Microcebus rufus / 122232)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Claire's mouse lemur (Microcebus mamiratra / 415757)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Goodman's mouse lemur (Microcebus lehilahytsara / 343908)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for gray mouse lemur (Microcebus murinus / 30608)\n",
      "\t105869284 -> GNRHR (Chromosome(s): ['26'], SwissProt: None)\n",
      "Fetch for Jolly's mouse lemur (Microcebus jollyae / 236275)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Margot Marsh's mouse lemur (Microcebus margotmarshae / 864581)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Microcebus bongolavensis / 412750)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Microcebus danfossi / 412751)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Microcebus griseorufus / 143354)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Microcebus myoxinus / 143283)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Microcebus ravelobensis / 122231)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Microcebus sambiranensis / 143353)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Microcebus tavaratra / 143351)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Mittermeier's mouse lemur (Microcebus mittermeieri / 236274)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Simmons' mouse lemur (Microcebus simmonsi / 236272)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Coquerel's mouse lemur (Mirza coquereli / 47180)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Mirza zaza / 339999)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Phaner electromontis / 1313324)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Phaner pallescens / 568393)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Phaner parienti / 1313325)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for woolly lemur (Avahi laniger / 122246)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Indri indri / 34827)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Coquerel's sifaka (Propithecus coquereli / 379532)\n",
      "\t105811193 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for diademed sifaka (Propithecus diadema / 83281)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Milne-Edwards sifaka (Propithecus edwardsi / 543559)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Tattersall's sifaka (Propithecus tattersalli / 30601)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for white sifaka (Propithecus verreauxi / 34825)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for black lemur (Eulemur macaco / 30602)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for brown lemur (Eulemur fulvus / 13515)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for mongoose lemur (Eulemur mongoz / 34828)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for red-bellied lemur (Eulemur rubriventer / 34829)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Red-fronted lemur (Eulemur rufus / 859983)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for bamboo lemur (Hapalemur griseus / 13557)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Ring-tailed lemur (Lemur catta / 9447)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for greater bamboo lemur (Prolemur simus / 1328070)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tGene-ids not found in this organism\n",
      "Fetch for ruffed lemur (Varecia variegata / 9455)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for gray-backed sportive lemur (Lepilemur dorsalis / 78583)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Hubbard's sportive lemur (Lepilemur hubbardorum / 756882)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for James' sportive lemur (Lepilemur jamesorum / 486960)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur aeeclis / 342399)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur ahmansonorum / 886957)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur ankaranensis / 342401)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur betsileo / 886958)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur fleuretae / 886959)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur grewcockorum / 886960)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur hollandorum / 886961)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur microdon / 185457)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur milanoii / 886963)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur mittermeieri / 378517)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur otto / 458173)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur petteri / 886964)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur sahamalazensis / 342398)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur scottorum / 886966)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur seali / 236288)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur tymerlachsoni / 886967)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Lepilemur wrightae / 886968)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Milne-Edwards's sportive lemur (Lepilemur edwardsi / 122230)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for northern sportive lemur (Lepilemur septentrionalis / 78584)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Randrianasolo's sportive lemur (Lepilemur randrianasoloi / 886965)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for red-tailed sportive lemur (Lepilemur ruficaudatus / 78866)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for white-footed sportive lemur (Lepilemur leucopus / 100475)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for None (Megaladapis edwardsi / 185953)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for large sloth lemur (Palaeopropithecus ingens / 1513477)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Moholi bushbaby (Galago moholi / 30609)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Senegal galago (Galago senegalensis / 9465)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for small-eared galago (Otolemur garnettii / 30611)\n",
      "\t100947835 -> GNRHR (Chromosome(s): ['Un'], SwissProt: None)\n",
      "Fetch for thick-tailed bush baby (Otolemur crassicaudatus / 9463)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for gray slender loris (Loris lydekkerianus / 300163)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for slender loris (Loris tardigradus / 9468)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for Bengal slow loris (Nycticebus bengalensis / 261741)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for pygmy slow loris (Nycticebus pygmaeus / 101278)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for slow loris (Nycticebus coucang / 9470)\n",
      "\tGene-ids not found in this organism\n",
      "Fetch for potto (Perodicticus potto / 9472)\n",
      "\tGene-ids not found in this organism\n"
     ]
    }
   ],
   "source": [
    "#Get gene descriptors by gene symbol + organism name for all primates\n",
    "primate_tax_id = 9443\n",
    "\n",
    "def species_tax_ids(tree):\n",
    "    \n",
    "    if tree.children:\n",
    "        for child in tree.children:\n",
    "            yield from species_tax_ids(child)\n",
    "        \n",
    "    if tree.rank == 'SPECIES':\n",
    "        yield tree.tax_id, tree.sci_name, tree.common_name\n",
    "\n",
    "primate_tax_tree = ds_gene_instance.gene_tax_tree(taxon=primate_tax_id)\n",
    "\n",
    "symbols = ['GNRHR', 'GNRHR2']\n",
    "for tax_id, sci_name, common_name in species_tax_ids(primate_tax_tree):\n",
    "    print(f'Fetch for {common_name} ({sci_name} / {tax_id})')\n",
    "    gene_descriptors = ds_gene_instance.gene_metadata_by_tax_and_symbol(symbols=symbols, taxon=tax_id)\n",
    "    report_on_gene_descriptors(gene_descriptors, report_errors=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Get gene descriptors for GNRHR genes across vertebrates and build a table\n",
    "\n",
    "Let's expand the taxonomic scope even further, and look at a selection of vertebrates.\n",
    "\n",
    "We'll use a pre-determined list of Gene IDs to get gene descriptors for these genes and build an easily readable table with key information about these genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>common_name</th>\n",
       "      <th>taxonomic_name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>num_transcripts</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>omim_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>nomenclature_id</th>\n",
       "      <th>nomenclature_auth</th>\n",
       "      <th>genome_coordinates</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>100001586</th>\n",
       "      <td>zebrafish</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>gnrhr4</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>[ENSDARG00000038116]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ZDB-GENE-050419-76</td>\n",
       "      <td>ZFIN</td>\n",
       "      <td>NC_007129.7:25390740_25402909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100135415</th>\n",
       "      <td>tropical clawed frog</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>gnrhr2</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>[ENSXETG00000021161]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>XB-GENE-5867415</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>NC_030679.2:109001475_109014876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100135416</th>\n",
       "      <td>tropical clawed frog</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>gnrhr2/nmi</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>[ENSXETG00000005637]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_030679.2:116455054_116472596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100135417</th>\n",
       "      <td>tropical clawed frog</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>gnrhr</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>[ENSXETG00000001290]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>XB-GENE-5753573</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>NC_030684.2:142203708_142211556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100136028</th>\n",
       "      <td>rainbow trout</td>\n",
       "      <td>Oncorhynchus mykiss</td>\n",
       "      <td>gnrh-r</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSOMYG00000000839]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_035078.1:14963448_14968063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100270671</th>\n",
       "      <td>zebrafish</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>gnrhr2</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSDARG00000003553]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ZDB-GENE-090128-3</td>\n",
       "      <td>ZFIN</td>\n",
       "      <td>NC_007118.7:52364748_52368843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100270672</th>\n",
       "      <td>zebrafish</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>gnrhr1</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>[ENSDARG00000100593]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ZDB-GENE-090128-2</td>\n",
       "      <td>ZFIN</td>\n",
       "      <td>NC_007130.7:43213293_43235554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101318246</th>\n",
       "      <td>common bottlenose dolphin</td>\n",
       "      <td>Tursiops truncatus</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_047038.1:85406171_85428905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101932446</th>\n",
       "      <td>Painted turtle</td>\n",
       "      <td>Chrysemys picta</td>\n",
       "      <td>LOC101932446</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSCPBG00000024942]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_007281386.1:1139258_1141674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101935915</th>\n",
       "      <td>Painted turtle</td>\n",
       "      <td>Chrysemys picta</td>\n",
       "      <td>LOC101935915</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>[ENSCPBG00000007948]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_007281382.1:370463_379501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101953943</th>\n",
       "      <td>Painted turtle</td>\n",
       "      <td>Chrysemys picta</td>\n",
       "      <td>LOC101953943</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>6</td>\n",
       "      <td>[ENSCPBG00000014672]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_007281446.1:3015207_3081999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102193667</th>\n",
       "      <td>None</td>\n",
       "      <td>Pundamilia nyererei</td>\n",
       "      <td>LOC102193667</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSPNYG00000019741]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_005187452.1:3461026_3466217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102202954</th>\n",
       "      <td>None</td>\n",
       "      <td>Pundamilia nyererei</td>\n",
       "      <td>gnrhr4</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_005187436.1:1835841_1849078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102205592</th>\n",
       "      <td>None</td>\n",
       "      <td>Pundamilia nyererei</td>\n",
       "      <td>LOC102205592</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSPNYG00000007579]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_005187426.1:621250_624817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102346610</th>\n",
       "      <td>coelacanth</td>\n",
       "      <td>Latimeria chalumnae</td>\n",
       "      <td>LOC102346610</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>[ENSLACG00000014236]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_005819113.1:988319_995279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102363373</th>\n",
       "      <td>coelacanth</td>\n",
       "      <td>Latimeria chalumnae</td>\n",
       "      <td>LOC102363373</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_005819068.1:1628819_1638066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102364206</th>\n",
       "      <td>coelacanth</td>\n",
       "      <td>Latimeria chalumnae</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSLACG00000013490]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_005819460.1:865222_884433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102366752</th>\n",
       "      <td>coelacanth</td>\n",
       "      <td>Latimeria chalumnae</td>\n",
       "      <td>GNRHR1/III</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_005819060.1:2885236_2899430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102536567</th>\n",
       "      <td>alpaca</td>\n",
       "      <td>Vicugna pacos</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_021964157.1:55397729_55411117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102687824</th>\n",
       "      <td>spotted gar</td>\n",
       "      <td>Lepisosteus oculatus</td>\n",
       "      <td>gnrhr4</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSLOCG00000014360]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_023181.1:47828941_47842900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102694185</th>\n",
       "      <td>spotted gar</td>\n",
       "      <td>Lepisosteus oculatus</td>\n",
       "      <td>LOC102694185</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>[ENSLOCG00000008760]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_023202.1:8297180_8319507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102770612</th>\n",
       "      <td>None</td>\n",
       "      <td>Myotis davidii</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_006283639.1:3309455_3329661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103899900</th>\n",
       "      <td>emperor penguin</td>\n",
       "      <td>Aptenodytes forsteri</td>\n",
       "      <td>LOC103899900</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_008795371.1:3421273_3430177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103899926</th>\n",
       "      <td>emperor penguin</td>\n",
       "      <td>Aptenodytes forsteri</td>\n",
       "      <td>LOC103899926</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_008795371.1:6612054_6613998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105916404</th>\n",
       "      <td>mummichog</td>\n",
       "      <td>Fundulus heteroclitus</td>\n",
       "      <td>LOC105916404</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSFHEG00000000340]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_012224575.1:3535642_3546823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105919697</th>\n",
       "      <td>mummichog</td>\n",
       "      <td>Fundulus heteroclitus</td>\n",
       "      <td>gnrhr4</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSFHEG00000022489]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_012224868.1:148392_172709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105934126</th>\n",
       "      <td>mummichog</td>\n",
       "      <td>Fundulus heteroclitus</td>\n",
       "      <td>LOC105934126</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSFHEG00000001414]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_012234394.1:1071748_1081214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108392639</th>\n",
       "      <td>Malayan pangolin</td>\n",
       "      <td>Manis javanica</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_016530059.1:16970_28019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109309182</th>\n",
       "      <td>Australian saltwater crocodile</td>\n",
       "      <td>Crocodylus porosus</td>\n",
       "      <td>LOC109309182</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>[ENSCPRG00005000240]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_017728911.1:36428760_36437582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109324103</th>\n",
       "      <td>Australian saltwater crocodile</td>\n",
       "      <td>Crocodylus porosus</td>\n",
       "      <td>LOC109324103</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSCPRG00005008895]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_017728893.1:1169521_1173474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109987527</th>\n",
       "      <td>ballan wrasse</td>\n",
       "      <td>Labrus bergylta</td>\n",
       "      <td>LOC109987527</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSLBEG00000028052]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_018114533.1:1324248_1334822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109994050</th>\n",
       "      <td>ballan wrasse</td>\n",
       "      <td>Labrus bergylta</td>\n",
       "      <td>LOC109994050</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_018114415.1:5852414_5857700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109999298</th>\n",
       "      <td>ballan wrasse</td>\n",
       "      <td>Labrus bergylta</td>\n",
       "      <td>gnrhr4</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSLBEG00000006912]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_018114964.1:235812_256320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110488224</th>\n",
       "      <td>rainbow trout</td>\n",
       "      <td>Oncorhynchus mykiss</td>\n",
       "      <td>LOC110488224</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSOMYG00000013327]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_035090.1:24287559_24322303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110495632</th>\n",
       "      <td>rainbow trout</td>\n",
       "      <td>Oncorhynchus mykiss</td>\n",
       "      <td>LOC110495632</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSOMYG00000026325]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_035078.1:56791316_56797819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110496352</th>\n",
       "      <td>rainbow trout</td>\n",
       "      <td>Oncorhynchus mykiss</td>\n",
       "      <td>LOC110496352</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSOMYG00000027791]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_035094.1:37625602_37635790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110513414</th>\n",
       "      <td>rainbow trout</td>\n",
       "      <td>Oncorhynchus mykiss</td>\n",
       "      <td>LOC110513414</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>[ENSOMYG00000000211]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_035079.1:24691038_24711154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110520912</th>\n",
       "      <td>rainbow trout</td>\n",
       "      <td>Oncorhynchus mykiss</td>\n",
       "      <td>gnrhr4</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSOMYG00000040422]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_035077.1:27314948_27342569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112994411</th>\n",
       "      <td>emu</td>\n",
       "      <td>Dromaius novaehollandiae</td>\n",
       "      <td>LOC112994411</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_020453730.1:6179_7726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112996301</th>\n",
       "      <td>emu</td>\n",
       "      <td>Dromaius novaehollandiae</td>\n",
       "      <td>LOC112996301</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSDNVG00000011124]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NW_020453948.1:1559259_1562466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114645297</th>\n",
       "      <td>reedfish</td>\n",
       "      <td>Erpetoichthys calabaricus</td>\n",
       "      <td>LOC114645297</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_041395.1:82016453_82036206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114667483</th>\n",
       "      <td>reedfish</td>\n",
       "      <td>Erpetoichthys calabaricus</td>\n",
       "      <td>gnrhr4</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSECRG00000016482]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_041410.1:98965514_99009309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114814</th>\n",
       "      <td>human</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GNRHR2</td>\n",
       "      <td>PSEUDO</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>[ENSG00000211451]</td>\n",
       "      <td>[612875]</td>\n",
       "      <td>[Q96P88.4]</td>\n",
       "      <td>HGNC:16341</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>NC_000001.11:145919013_145925341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14715</th>\n",
       "      <td>house mouse</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Gnrhr</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>4</td>\n",
       "      <td>[ENSMUSG00000029255]</td>\n",
       "      <td>None</td>\n",
       "      <td>[Q01776.1]</td>\n",
       "      <td>MGI:95790</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NC_000071.6:86180754_86197901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2798</th>\n",
       "      <td>human</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>[ENSG00000109163]</td>\n",
       "      <td>[138850]</td>\n",
       "      <td>[P30968.1]</td>\n",
       "      <td>HGNC:4421</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>NC_000004.12:67737118_67754388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281798</th>\n",
       "      <td>cattle</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>[ENSBTAG00000000438]</td>\n",
       "      <td>None</td>\n",
       "      <td>[P32236.1]</td>\n",
       "      <td>VGNC:29482</td>\n",
       "      <td>VGNC</td>\n",
       "      <td>NC_037333.1:83434759_83452222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395368</th>\n",
       "      <td>chicken</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>[ENSGALG00000021020]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>49286</td>\n",
       "      <td>CGNC</td>\n",
       "      <td>NC_006097.5:18239128_18254320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403718</th>\n",
       "      <td>dog</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSCAFG00000002785]</td>\n",
       "      <td>None</td>\n",
       "      <td>[Q9MZI6.1]</td>\n",
       "      <td>VGNC:41337</td>\n",
       "      <td>VGNC</td>\n",
       "      <td>NC_006595.3:58253112_58266600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404718</th>\n",
       "      <td>human</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GNRHR2P1</td>\n",
       "      <td>PSEUDO</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>[ENSG00000259169]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:39424</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>NC_000014.9:60398823_60400013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427517</th>\n",
       "      <td>chicken</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>[ENSGALG00000020923]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>53110</td>\n",
       "      <td>CGNC</td>\n",
       "      <td>NC_006097.5:20425797_20428105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471226</th>\n",
       "      <td>chimpanzee</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>PROTEIN_CODING</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>[ENSPTRG00000016100]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>VGNC:2019</td>\n",
       "      <td>VGNC</td>\n",
       "      <td>NC_036883.1:61240135_61260886</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              common_name             taxonomic_name  \\\n",
       "gene_id                                                                \n",
       "100001586                       zebrafish                Danio rerio   \n",
       "100135415            tropical clawed frog         Xenopus tropicalis   \n",
       "100135416            tropical clawed frog         Xenopus tropicalis   \n",
       "100135417            tropical clawed frog         Xenopus tropicalis   \n",
       "100136028                   rainbow trout        Oncorhynchus mykiss   \n",
       "100270671                       zebrafish                Danio rerio   \n",
       "100270672                       zebrafish                Danio rerio   \n",
       "101318246       common bottlenose dolphin         Tursiops truncatus   \n",
       "101932446                  Painted turtle            Chrysemys picta   \n",
       "101935915                  Painted turtle            Chrysemys picta   \n",
       "101953943                  Painted turtle            Chrysemys picta   \n",
       "102193667                            None        Pundamilia nyererei   \n",
       "102202954                            None        Pundamilia nyererei   \n",
       "102205592                            None        Pundamilia nyererei   \n",
       "102346610                      coelacanth        Latimeria chalumnae   \n",
       "102363373                      coelacanth        Latimeria chalumnae   \n",
       "102364206                      coelacanth        Latimeria chalumnae   \n",
       "102366752                      coelacanth        Latimeria chalumnae   \n",
       "102536567                          alpaca              Vicugna pacos   \n",
       "102687824                     spotted gar       Lepisosteus oculatus   \n",
       "102694185                     spotted gar       Lepisosteus oculatus   \n",
       "102770612                            None             Myotis davidii   \n",
       "103899900                 emperor penguin       Aptenodytes forsteri   \n",
       "103899926                 emperor penguin       Aptenodytes forsteri   \n",
       "105916404                       mummichog      Fundulus heteroclitus   \n",
       "105919697                       mummichog      Fundulus heteroclitus   \n",
       "105934126                       mummichog      Fundulus heteroclitus   \n",
       "108392639                Malayan pangolin             Manis javanica   \n",
       "109309182  Australian saltwater crocodile         Crocodylus porosus   \n",
       "109324103  Australian saltwater crocodile         Crocodylus porosus   \n",
       "109987527                   ballan wrasse            Labrus bergylta   \n",
       "109994050                   ballan wrasse            Labrus bergylta   \n",
       "109999298                   ballan wrasse            Labrus bergylta   \n",
       "110488224                   rainbow trout        Oncorhynchus mykiss   \n",
       "110495632                   rainbow trout        Oncorhynchus mykiss   \n",
       "110496352                   rainbow trout        Oncorhynchus mykiss   \n",
       "110513414                   rainbow trout        Oncorhynchus mykiss   \n",
       "110520912                   rainbow trout        Oncorhynchus mykiss   \n",
       "112994411                             emu   Dromaius novaehollandiae   \n",
       "112996301                             emu   Dromaius novaehollandiae   \n",
       "114645297                        reedfish  Erpetoichthys calabaricus   \n",
       "114667483                        reedfish  Erpetoichthys calabaricus   \n",
       "114814                              human               Homo sapiens   \n",
       "14715                         house mouse               Mus musculus   \n",
       "2798                                human               Homo sapiens   \n",
       "281798                             cattle                 Bos taurus   \n",
       "395368                            chicken              Gallus gallus   \n",
       "403718                                dog     Canis lupus familiaris   \n",
       "404718                              human               Homo sapiens   \n",
       "427517                            chicken              Gallus gallus   \n",
       "471226                         chimpanzee            Pan troglodytes   \n",
       "\n",
       "                 symbol            type chromosome  num_transcripts  \\\n",
       "gene_id                                                               \n",
       "100001586        gnrhr4  PROTEIN_CODING       None                2   \n",
       "100135415        gnrhr2  PROTEIN_CODING       None                3   \n",
       "100135416    gnrhr2/nmi  PROTEIN_CODING       None                2   \n",
       "100135417         gnrhr  PROTEIN_CODING       None                2   \n",
       "100136028        gnrh-r  PROTEIN_CODING       None                1   \n",
       "100270671        gnrhr2  PROTEIN_CODING       None                1   \n",
       "100270672        gnrhr1  PROTEIN_CODING       None                2   \n",
       "101318246         GNRHR  PROTEIN_CODING       None                1   \n",
       "101932446  LOC101932446  PROTEIN_CODING       None                1   \n",
       "101935915  LOC101935915  PROTEIN_CODING       None                2   \n",
       "101953943  LOC101953943  PROTEIN_CODING       None                6   \n",
       "102193667  LOC102193667  PROTEIN_CODING       None                1   \n",
       "102202954        gnrhr4  PROTEIN_CODING       None                1   \n",
       "102205592  LOC102205592  PROTEIN_CODING       None                1   \n",
       "102346610  LOC102346610  PROTEIN_CODING       None                2   \n",
       "102363373  LOC102363373  PROTEIN_CODING       None                1   \n",
       "102364206         GNRHR  PROTEIN_CODING       None                1   \n",
       "102366752    GNRHR1/III  PROTEIN_CODING       None                1   \n",
       "102536567         GNRHR  PROTEIN_CODING       None                2   \n",
       "102687824        gnrhr4  PROTEIN_CODING       None                1   \n",
       "102694185  LOC102694185  PROTEIN_CODING       None                3   \n",
       "102770612         GNRHR  PROTEIN_CODING       None                1   \n",
       "103899900  LOC103899900  PROTEIN_CODING       None                1   \n",
       "103899926  LOC103899926  PROTEIN_CODING       None                1   \n",
       "105916404  LOC105916404  PROTEIN_CODING       None                1   \n",
       "105919697        gnrhr4  PROTEIN_CODING       None                1   \n",
       "105934126  LOC105934126  PROTEIN_CODING       None                1   \n",
       "108392639         GNRHR  PROTEIN_CODING       None                2   \n",
       "109309182  LOC109309182  PROTEIN_CODING       None                2   \n",
       "109324103  LOC109324103  PROTEIN_CODING       None                1   \n",
       "109987527  LOC109987527  PROTEIN_CODING       None                1   \n",
       "109994050  LOC109994050  PROTEIN_CODING       None                1   \n",
       "109999298        gnrhr4  PROTEIN_CODING       None                1   \n",
       "110488224  LOC110488224  PROTEIN_CODING       None                1   \n",
       "110495632  LOC110495632  PROTEIN_CODING       None                1   \n",
       "110496352  LOC110496352  PROTEIN_CODING       None                1   \n",
       "110513414  LOC110513414  PROTEIN_CODING       None                2   \n",
       "110520912        gnrhr4  PROTEIN_CODING       None                1   \n",
       "112994411  LOC112994411  PROTEIN_CODING       None                1   \n",
       "112996301  LOC112996301  PROTEIN_CODING       None                1   \n",
       "114645297  LOC114645297  PROTEIN_CODING       None                1   \n",
       "114667483        gnrhr4  PROTEIN_CODING       None                1   \n",
       "114814           GNRHR2          PSEUDO       None                3   \n",
       "14715             Gnrhr  PROTEIN_CODING       None                4   \n",
       "2798              GNRHR  PROTEIN_CODING       None                2   \n",
       "281798            GNRHR  PROTEIN_CODING       None                2   \n",
       "395368            GNRHR  PROTEIN_CODING       None                3   \n",
       "403718            GNRHR  PROTEIN_CODING       None                1   \n",
       "404718         GNRHR2P1          PSEUDO       None                0   \n",
       "427517            GNRHR  PROTEIN_CODING       None                1   \n",
       "471226            GNRHR  PROTEIN_CODING       None                2   \n",
       "\n",
       "                     ensembl_id   omim_id  uniprot_id     nomenclature_id  \\\n",
       "gene_id                                                                     \n",
       "100001586  [ENSDARG00000038116]      None        None  ZDB-GENE-050419-76   \n",
       "100135415  [ENSXETG00000021161]      None        None     XB-GENE-5867415   \n",
       "100135416  [ENSXETG00000005637]      None        None                None   \n",
       "100135417  [ENSXETG00000001290]      None        None     XB-GENE-5753573   \n",
       "100136028  [ENSOMYG00000000839]      None        None                None   \n",
       "100270671  [ENSDARG00000003553]      None        None   ZDB-GENE-090128-3   \n",
       "100270672  [ENSDARG00000100593]      None        None   ZDB-GENE-090128-2   \n",
       "101318246                  None      None        None                None   \n",
       "101932446  [ENSCPBG00000024942]      None        None                None   \n",
       "101935915  [ENSCPBG00000007948]      None        None                None   \n",
       "101953943  [ENSCPBG00000014672]      None        None                None   \n",
       "102193667  [ENSPNYG00000019741]      None        None                None   \n",
       "102202954                  None      None        None                None   \n",
       "102205592  [ENSPNYG00000007579]      None        None                None   \n",
       "102346610  [ENSLACG00000014236]      None        None                None   \n",
       "102363373                  None      None        None                None   \n",
       "102364206  [ENSLACG00000013490]      None        None                None   \n",
       "102366752                  None      None        None                None   \n",
       "102536567                  None      None        None                None   \n",
       "102687824  [ENSLOCG00000014360]      None        None                None   \n",
       "102694185  [ENSLOCG00000008760]      None        None                None   \n",
       "102770612                  None      None        None                None   \n",
       "103899900                  None      None        None                None   \n",
       "103899926                  None      None        None                None   \n",
       "105916404  [ENSFHEG00000000340]      None        None                None   \n",
       "105919697  [ENSFHEG00000022489]      None        None                None   \n",
       "105934126  [ENSFHEG00000001414]      None        None                None   \n",
       "108392639                  None      None        None                None   \n",
       "109309182  [ENSCPRG00005000240]      None        None                None   \n",
       "109324103  [ENSCPRG00005008895]      None        None                None   \n",
       "109987527  [ENSLBEG00000028052]      None        None                None   \n",
       "109994050                  None      None        None                None   \n",
       "109999298  [ENSLBEG00000006912]      None        None                None   \n",
       "110488224  [ENSOMYG00000013327]      None        None                None   \n",
       "110495632  [ENSOMYG00000026325]      None        None                None   \n",
       "110496352  [ENSOMYG00000027791]      None        None                None   \n",
       "110513414  [ENSOMYG00000000211]      None        None                None   \n",
       "110520912  [ENSOMYG00000040422]      None        None                None   \n",
       "112994411                  None      None        None                None   \n",
       "112996301  [ENSDNVG00000011124]      None        None                None   \n",
       "114645297                  None      None        None                None   \n",
       "114667483  [ENSECRG00000016482]      None        None                None   \n",
       "114814        [ENSG00000211451]  [612875]  [Q96P88.4]          HGNC:16341   \n",
       "14715      [ENSMUSG00000029255]      None  [Q01776.1]           MGI:95790   \n",
       "2798          [ENSG00000109163]  [138850]  [P30968.1]           HGNC:4421   \n",
       "281798     [ENSBTAG00000000438]      None  [P32236.1]          VGNC:29482   \n",
       "395368     [ENSGALG00000021020]      None        None               49286   \n",
       "403718     [ENSCAFG00000002785]      None  [Q9MZI6.1]          VGNC:41337   \n",
       "404718        [ENSG00000259169]      None        None          HGNC:39424   \n",
       "427517     [ENSGALG00000020923]      None        None               53110   \n",
       "471226     [ENSPTRG00000016100]      None        None           VGNC:2019   \n",
       "\n",
       "          nomenclature_auth                genome_coordinates  \n",
       "gene_id                                                        \n",
       "100001586              ZFIN     NC_007129.7:25390740_25402909  \n",
       "100135415           Xenbase   NC_030679.2:109001475_109014876  \n",
       "100135416              None   NC_030679.2:116455054_116472596  \n",
       "100135417           Xenbase   NC_030684.2:142203708_142211556  \n",
       "100136028              None     NC_035078.1:14963448_14968063  \n",
       "100270671              ZFIN     NC_007118.7:52364748_52368843  \n",
       "100270672              ZFIN     NC_007130.7:43213293_43235554  \n",
       "101318246              None     NC_047038.1:85406171_85428905  \n",
       "101932446              None    NW_007281386.1:1139258_1141674  \n",
       "101935915              None      NW_007281382.1:370463_379501  \n",
       "101953943              None    NW_007281446.1:3015207_3081999  \n",
       "102193667              None    NW_005187452.1:3461026_3466217  \n",
       "102202954              None    NW_005187436.1:1835841_1849078  \n",
       "102205592              None      NW_005187426.1:621250_624817  \n",
       "102346610              None      NW_005819113.1:988319_995279  \n",
       "102363373              None    NW_005819068.1:1628819_1638066  \n",
       "102364206              None      NW_005819460.1:865222_884433  \n",
       "102366752              None    NW_005819060.1:2885236_2899430  \n",
       "102536567              None  NW_021964157.1:55397729_55411117  \n",
       "102687824              None     NC_023181.1:47828941_47842900  \n",
       "102694185              None       NC_023202.1:8297180_8319507  \n",
       "102770612              None    NW_006283639.1:3309455_3329661  \n",
       "103899900              None    NW_008795371.1:3421273_3430177  \n",
       "103899926              None    NW_008795371.1:6612054_6613998  \n",
       "105916404              None    NW_012224575.1:3535642_3546823  \n",
       "105919697              None      NW_012224868.1:148392_172709  \n",
       "105934126              None    NW_012234394.1:1071748_1081214  \n",
       "108392639              None        NW_016530059.1:16970_28019  \n",
       "109309182              None  NW_017728911.1:36428760_36437582  \n",
       "109324103              None    NW_017728893.1:1169521_1173474  \n",
       "109987527              None    NW_018114533.1:1324248_1334822  \n",
       "109994050              None    NW_018114415.1:5852414_5857700  \n",
       "109999298              None      NW_018114964.1:235812_256320  \n",
       "110488224              None     NC_035090.1:24287559_24322303  \n",
       "110495632              None     NC_035078.1:56791316_56797819  \n",
       "110496352              None     NC_035094.1:37625602_37635790  \n",
       "110513414              None     NC_035079.1:24691038_24711154  \n",
       "110520912              None     NC_035077.1:27314948_27342569  \n",
       "112994411              None          NW_020453730.1:6179_7726  \n",
       "112996301              None    NW_020453948.1:1559259_1562466  \n",
       "114645297              None     NC_041395.1:82016453_82036206  \n",
       "114667483              None     NC_041410.1:98965514_99009309  \n",
       "114814                 HGNC  NC_000001.11:145919013_145925341  \n",
       "14715                   MGI     NC_000071.6:86180754_86197901  \n",
       "2798                   HGNC    NC_000004.12:67737118_67754388  \n",
       "281798                 VGNC     NC_037333.1:83434759_83452222  \n",
       "395368                 CGNC     NC_006097.5:18239128_18254320  \n",
       "403718                 VGNC     NC_006595.3:58253112_58266600  \n",
       "404718                 HGNC     NC_000014.9:60398823_60400013  \n",
       "427517                 CGNC     NC_006097.5:20425797_20428105  \n",
       "471226                 VGNC     NC_036883.1:61240135_61260886  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract fields of interest from descriptors class to build a table\n",
    "cols = '''\n",
    "common_name\n",
    "taxonomic_name\n",
    "symbol\n",
    "type\n",
    "chromosome\n",
    "num_transcripts\n",
    "ensembl_id\n",
    "omim_id\n",
    "uniprot_id\n",
    "nomenclature_id\n",
    "nomenclature_auth\n",
    "genome_coordinates\n",
    "'''\n",
    "cols = cols.split('\\n')[1:-1]\n",
    "\n",
    "def _range_repr(range):\n",
    "    ret = []\n",
    "    for interval in range:\n",
    "        ret.append(f'{interval.begin}_{interval.end}')\n",
    "    return ','.join(ret)\n",
    "\n",
    "def _ranges_repr(ranges):\n",
    "    ret = []\n",
    "    for range in ranges:\n",
    "        ret.append(f'{range.accession_version}:{_range_repr(range.range)}')\n",
    "    return ','.join(ret)\n",
    "\n",
    "# specify genes of interest and retrieve descriptors\n",
    "gene_ids = [2798, 114814, 404718, 14715, 109324103, 109309182, 281798, 395368, 403718, 427517, 471226, 7226731, 100001586, 100135415, 100135416, 100135417, 100136028, 100270671, 100270672, 101318246, 101932446, 101935915, 101953943, 102193667, 102202954, 102205592, 102346610, 102363373, 102364206, 102366752, 102536567, 102687824, 102694185, 102770612, 103899900, 103899926, 105916404, 105919697, 105934126, 108392639, 109987527, 109994050, 109999298, 110488224, 110495632, 110496352, 110513414, 110520912, 112994411, 112996301, 114645297, 114667483]\n",
    "gene_descriptors = ds_gene_instance.gene_metadata_by_id(gene_ids)\n",
    "\n",
    "# collect elements of the descriptor class into a dictionary based on each gene ID\n",
    "table_data = {}\n",
    "for gene in map(lambda g: g.gene, gene_descriptors.genes):\n",
    "    table_data[gene.gene_id] = [gene.common_name]\n",
    "    table_data[gene.gene_id].append(gene.taxname)\n",
    "    table_data[gene.gene_id].append(gene.symbol)\n",
    "    table_data[gene.gene_id].append(gene.type)\n",
    "    table_data[gene.gene_id].append(gene.chromosome)\n",
    "    if gene.transcripts:\n",
    "        table_data[gene.gene_id].append(len(gene.transcripts))\n",
    "    else:\n",
    "        table_data[gene.gene_id].append(0)\n",
    "    table_data[gene.gene_id].append(gene.ensembl_gene_ids)\n",
    "    table_data[gene.gene_id].append(gene.omim_ids)\n",
    "    table_data[gene.gene_id].append(gene.swiss_prot_accessions)\n",
    "    if gene.nomenclature_authority:\n",
    "        table_data[gene.gene_id].append(gene.nomenclature_authority.identifier)\n",
    "        table_data[gene.gene_id].append(gene.nomenclature_authority.authority)\n",
    "    else:\n",
    "        table_data[gene.gene_id].append(None)\n",
    "        table_data[gene.gene_id].append(None)        \n",
    "    table_data[gene.gene_id].append(_ranges_repr(gene.genomic_ranges))\n",
    "\n",
    "        \n",
    "df = pd.DataFrame.from_dict(table_data, orient='index', columns=cols)\n",
    "df.index.name = 'gene_id'\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Use gene descriptor data to build a simple table of gene copy numbers\n",
    "\n",
    "As we mentioned before, while the role of the gonadotropin releasing hormone and receptor in reproduction is highly conserved across vertebrates, gene copy number for the receptor genes is not. Let's take a quick look at how gene count varies in different organisms.  \n",
    "Note that in our list, you can see that rainbow trout has the most gene copies at 6, while numerous mammals, including mouse, dolphin and alpaca, only have a single annotated gene copy. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rainbow trout</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>coelacanth</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>zebrafish</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Painted turtle</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tropical clawed frog</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ballan wrasse</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>mummichog</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>human</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>emu</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>spotted gar</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>reedfish</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Australian saltwater crocodile</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>emperor penguin</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chicken</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>house mouse</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Malayan pangolin</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>common bottlenose dolphin</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chimpanzee</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cattle</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>alpaca</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>dog</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          organism  gene_count\n",
       "16                   rainbow trout           6\n",
       "8                       coelacanth           4\n",
       "20                       zebrafish           3\n",
       "2                   Painted turtle           3\n",
       "19            tropical clawed frog           3\n",
       "4                    ballan wrasse           3\n",
       "15                       mummichog           3\n",
       "14                           human           3\n",
       "12                             emu           2\n",
       "18                     spotted gar           2\n",
       "17                        reedfish           2\n",
       "0   Australian saltwater crocodile           2\n",
       "11                 emperor penguin           2\n",
       "6                          chicken           2\n",
       "13                     house mouse           1\n",
       "1                 Malayan pangolin           1\n",
       "9        common bottlenose dolphin           1\n",
       "7                       chimpanzee           1\n",
       "5                           cattle           1\n",
       "3                           alpaca           1\n",
       "10                             dog           1"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# plot gene count based on organism\n",
    "gene_cnt = df.groupby('common_name')['symbol'].count().reset_index()\n",
    "gene_cnt.columns = ['organism', 'gene_count']\n",
    "gene_cnt.sort_values('gene_count', ascending=False, inplace=True)\n",
    "gene_cnt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Use gene datasets to build a transcript-focused table\n",
    "\n",
    "Finally, we are going to download a gene dataset for the human GnRH receptor genes, and use metadata included within the dataset to build a transcript-focused table.\n",
    "We'll take this metadata from the data report file, which is similar to the gene descriptor described above, but in a more human-readable yaml format.  \n",
    "Note that gene datasets include a data table that is gene-oriented, with one gene record per row. Here, we are building a custom transcript-oriented table.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Download complete to gene_ds.zip\n",
      "CPU times: user 13 ms, sys: 3.39 ms, total: 16.4 ms\n",
      "Wall time: 176 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "gene_ids = [2798, 114814, 404718]\n",
    "gene_ds_download = ds_gene_instance.download_gene_package(gene_ids, _preload_content=False)\n",
    "\n",
    "## write to a zip file \n",
    "zipfile_name = 'gene_ds.zip'\n",
    "with open(zipfile_name, 'wb') as f:\n",
    "    f.write(gene_ds_download.data)\n",
    "\n",
    "print(f'Download complete to {zipfile_name}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "slideshow": {
     "slide_type": "notes"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Archive:  gene_ds.zip\r\n",
      " Length   Method    Size  Cmpr    Date    Time   CRC-32   Name\r\n",
      "--------  ------  ------- ---- ---------- ----- --------  ----\r\n",
      "     661  Defl:N      384  42% 09-04-2020 16:27 bc3c97af  README.md\r\n",
      "    5515  Defl:N     1284  77% 09-04-2020 16:27 c33d83dc  ncbi_dataset/data/data_report.yaml\r\n",
      "    1434  Defl:N      478  67% 09-04-2020 16:27 fa82a51a  ncbi_dataset/data/data_table.tsv\r\n",
      "     202  Defl:N      115  43% 09-04-2020 16:27 c98a95f2  ncbi_dataset/data/dataset_catalog.json\r\n",
      "       0  Defl:N        2   0% 09-04-2020 16:27 00000000  ncbi_dataset/fetch.txt\r\n",
      "--------          -------  ---                            -------\r\n",
      "    7812             2263  71%                            5 files\r\n"
     ]
    }
   ],
   "source": [
    "!unzip -v {zipfile_name}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_taxonomy</th>\n",
       "      <th>accVer</th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>5`UTR_len</th>\n",
       "      <th>3`UTR_len</th>\n",
       "      <th>protAccVer</th>\n",
       "      <th>protName</th>\n",
       "      <th>protLength</th>\n",
       "      <th>exonAccVer</th>\n",
       "      <th>numExons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>114814</td>\n",
       "      <td>GNRHR2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NR_002328.4</td>\n",
       "      <td>transcript variant 1</td>\n",
       "      <td>1626</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>114814</td>\n",
       "      <td>GNRHR2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NR_104034.1</td>\n",
       "      <td>transcript variant 3</td>\n",
       "      <td>786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>114814</td>\n",
       "      <td>GNRHR2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NR_104033.1</td>\n",
       "      <td>transcript variant 2</td>\n",
       "      <td>1035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2798</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NM_000406.3</td>\n",
       "      <td>transcript variant 1</td>\n",
       "      <td>4402</td>\n",
       "      <td>53.0</td>\n",
       "      <td>3362.0</td>\n",
       "      <td>NP_000397.1</td>\n",
       "      <td>isoform 1</td>\n",
       "      <td>328.0</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2798</td>\n",
       "      <td>GNRHR</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NM_001012763.2</td>\n",
       "      <td>transcript variant 2</td>\n",
       "      <td>4017</td>\n",
       "      <td>53.0</td>\n",
       "      <td>3214.0</td>\n",
       "      <td>NP_001012781.1</td>\n",
       "      <td>isoform 2</td>\n",
       "      <td>249.0</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id gene_symbol gene_taxonomy          accVer                  name  \\\n",
       "0   114814      GNRHR2  Homo sapiens     NR_002328.4  transcript variant 1   \n",
       "1   114814      GNRHR2  Homo sapiens     NR_104034.1  transcript variant 3   \n",
       "2   114814      GNRHR2  Homo sapiens     NR_104033.1  transcript variant 2   \n",
       "3     2798       GNRHR  Homo sapiens     NM_000406.3  transcript variant 1   \n",
       "4     2798       GNRHR  Homo sapiens  NM_001012763.2  transcript variant 2   \n",
       "\n",
       "   length  5`UTR_len  3`UTR_len      protAccVer   protName  protLength  \\\n",
       "0    1626        NaN        NaN            None       None         NaN   \n",
       "1     786        NaN        NaN            None       None         NaN   \n",
       "2    1035        NaN        NaN            None       None         NaN   \n",
       "3    4402       53.0     3362.0     NP_000397.1  isoform 1       328.0   \n",
       "4    4017       53.0     3214.0  NP_001012781.1  isoform 2       249.0   \n",
       "\n",
       "     exonAccVer  numExons  \n",
       "0  NC_000001.11         3  \n",
       "1  NC_000001.11         3  \n",
       "2  NC_000001.11         4  \n",
       "3  NC_000004.12         3  \n",
       "4  NC_000004.12         3  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import json\n",
    "import zipfile\n",
    "\n",
    "from google.protobuf.json_format import ParseDict\n",
    "import yaml\n",
    "\n",
    "import ncbi.datasets.v1alpha1.reports.gene_pb2 as gene_report_pb2\n",
    "\n",
    "def gene_report_for(path_to_zipfile):\n",
    "    '''\n",
    "    Return an object representing the data report.\n",
    "    path_to_zipfile: The relative path to the zipfile containing the virus data report\n",
    "    '''\n",
    "    with zipfile.ZipFile(path_to_zipfile, 'r') as zip:\n",
    "        gene_report_as_dict = yaml.safe_load(zip.read('ncbi_dataset/data/data_report.yaml'))\n",
    "    gene_report = gene_report_pb2.GeneDescriptors()\n",
    "    ParseDict(gene_report_as_dict, gene_report)\n",
    "    return gene_report\n",
    "\n",
    "def _5prime_len(transcript):\n",
    "    if not transcript.cds or not transcript.cds.range:\n",
    "        return None\n",
    "    return transcript.cds.range[0].begin - 1\n",
    "\n",
    "def _3prime_len(transcript):\n",
    "    if not transcript.cds or not transcript.cds.range:\n",
    "        return None\n",
    "    return transcript.length - transcript.cds.range[0].end\n",
    "\n",
    "gene_report = gene_report_for(zipfile_name)\n",
    "\n",
    "rows = []\n",
    "for gene in gene_report.genes:\n",
    "\n",
    "    # transcripts for each gene are embedded as lists and require additional handling\n",
    "    for transcript in gene.transcripts:\n",
    "        rows.append({\n",
    "            'gene_id': gene.gene_id,\n",
    "            'gene_symbol': gene.symbol,\n",
    "            'gene_taxonomy': gene.taxname,            \n",
    "            'accVer': transcript.accession_version,\n",
    "            'name': transcript.name,\n",
    "            'length': transcript.length,\n",
    "            '5`UTR_len': _5prime_len(transcript),\n",
    "            '3`UTR_len': _3prime_len(transcript),\n",
    "            'protAccVer': transcript.protein.accession_version or None,\n",
    "            'protName': transcript.protein.isoform_name or None,\n",
    "            'protLength': transcript.protein.length or None,\n",
    "            'exonAccVer': transcript.exons.accession_version,\n",
    "            'numExons': len(transcript.exons.range),\n",
    "        })\n",
    "\n",
    "transcript_table = pd.DataFrame(rows)\n",
    "\n",
    "transcript_table\n"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Slideshow",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
