{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using `ncbi.datasets` library to download and parse genome annotation data in GFF3 format\n",
    "\n",
    "The objective of this notebook is to use the `ncbi.datasets` python library to demonstrate how to download and parse genome annotation data in gff3 format from NCBI Datasets.\n",
    "\n",
    "There are two major types of genome data available from NCBI Datasets:\n",
    "\n",
    "1. Genome datasets, which include genome, transcript, and protein sequences, genome annotation data in gff3 format, an assembly data report (genome assembly and annotation metadata) and a sequence report (a list of sequences that comprise the genome assembly).  \n",
    "\n",
    "2. Genome summaries, which are sets of key metadata that describe the genome datasets \n",
    "\n",
    "As an example, we will download gff3 files for five _Lactobacillus_ genome assemblies, then parse those gff3 files to get information about crispr gene order in the corresponding genome assemblies.\n",
    "\n",
    "The notebook can be broken down into the following four major tasks:\n",
    "1. Query NCBI for _Lactobacillus_ assemblies and download the genome summaries for all available genome assemblies\n",
    "2. Parse the genome summaries to make a list of genome assemblies annotated in 2020\n",
    "3. Download sequence and annotation (gff3 and protein fasta) for five of those assemblies\n",
    "4. Parse the gff3 files to look at crispr gene order in those assemblies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "After importing the various python modules we will need, create the assembly and download API instances."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict, Counter\n",
    "from datetime import datetime\n",
    "import json\n",
    "import os\n",
    "from textwrap import dedent\n",
    "import zipfile\n",
    "\n",
    "import gffutils\n",
    "import matplotlib.pyplot as plt\n",
    "import ncbi.datasets\n",
    "import pandas as pd\n",
    "from pyfaidx import Fasta\n",
    "\n",
    "plt.style.use('ggplot')\n",
    "\n",
    "genome_api = ncbi.datasets.GenomeApi(ncbi.datasets.ApiClient())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Group genome assemblies based on annotation date\n",
    "Let's look at RefSeq *lactobacillus* genome assemblies, and see how we can group them based on annotation date. Annotation date, when available, is described in the genome summary.\n",
    "\n",
    "This section will include the following two major tasks:\n",
    "1. Query NCBI for Lactobacillus assemblies and download the genome summaries for all available genome assemblies\n",
    "2. Make a list of genome assemblies annotated in 2020\n",
    "\n",
    "Using the `ncbi.datasets` library, let's first get the count of available RefSeq *lactobacillus* assemblies. By setting `limit='none'`, we tell the API to return a genome summary with only the count of genome assemblies for the specified NCBI Taxonomy ID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of RefSeq lactobacillus genome assemblies: 1060\n"
     ]
    }
   ],
   "source": [
    "taxid = '1578' ## This is the NCBI Taxonomy ID for lactobacillus\n",
    "genome_summary = genome_api.assembly_descriptors_by_taxon(\n",
    "    taxon=taxid,\n",
    "    page_size=1,\n",
    "    filters_assembly_source='refseq')\n",
    "\n",
    "print(f\"Number of RefSeq lactobacillus genome assemblies: {genome_summary.total_count}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's download the genome summaries for these genome assemblies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome summaries for 1060 genome assemblies downloaded and saved to genome_summary.\n",
      "As an example, here's the first genome summary for the first genome assembly in the list:\n",
      " {'assembly': {'annotation_metadata': {'file': [{'estimated_size': '111943',\n",
      "                                                'type': 'GENOME_GFF'},\n",
      "                                               {'estimated_size': '1187873',\n",
      "                                                'type': 'GENOME_GBFF'},\n",
      "                                               {'estimated_size': '301643',\n",
      "                                                'type': 'PROT_FASTA'},\n",
      "                                               {'estimated_size': '126907',\n",
      "                                                'type': 'GENOME_GTF'}],\n",
      "                                      'name': 'From INSDC submitter',\n",
      "                                      'release_date': '2019-10-07',\n",
      "                                      'source': 'Korea University',\n",
      "                                      'stats': {'gene_counts': {'protein_coding': 1517,\n",
      "                                                                'total': 1646}}},\n",
      "              'assembly_accession': 'GCF_008831485.1',\n",
      "              'assembly_category': 'representative genome',\n",
      "              'assembly_level': 'Complete Genome',\n",
      "              'bioproject_lineages': [{'bioprojects': [{'accession': 'PRJNA566216',\n",
      "                                                        'title': 'Lactobacillus '\n",
      "                                                                 'acetotolerans '\n",
      "                                                                 'Genome '\n",
      "                                                                 'sequencing '\n",
      "                                                                 'and '\n",
      "                                                                 'assembly'}]}],\n",
      "              'biosample_accession': 'SAMN12784818',\n",
      "              'chromosomes': [{'accession_version': 'NZ_CP044496.1',\n",
      "                               'length': '1683905',\n",
      "                               'name': 'ANONYMOUS'}],\n",
      "              'contig_n50': 1683905,\n",
      "              'display_name': 'ASM883148v1',\n",
      "              'estimated_size': '2223244',\n",
      "              'org': {'assembly_counts': {'node': 12, 'subtree': 16},\n",
      "                      'key': '1600',\n",
      "                      'parent_tax_id': '1578',\n",
      "                      'rank': 'SPECIES',\n",
      "                      'sci_name': 'Lactobacillus acetotolerans',\n",
      "                      'strain': 'LA749',\n",
      "                      'tax_id': '1600',\n",
      "                      'title': 'Lactobacillus acetotolerans'},\n",
      "              'paired_assembly_accession': 'GCA_008831485.1',\n",
      "              'seq_length': '1683905',\n",
      "              'submission_date': '2019-10-07',\n",
      "              'submitter': 'Korea University'}}\n"
     ]
    }
   ],
   "source": [
    "## download the genome summaries for all RefSeq lactobacillus genome assemblies\n",
    "genome_summary = genome_api.assembly_descriptors_by_taxon(\n",
    "    taxon=taxid,\n",
    "    page_size=1000,\n",
    "    filters_assembly_source='refseq')\n",
    "print(f\"Genome summaries for {genome_summary.total_count} genome assemblies downloaded and saved to genome_summary.\")\n",
    "print(f\"As an example, here's the first genome summary for the first genome assembly in the list:\\n\", genome_summary.assemblies[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we'll plot the assemblies by the year they were annotated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "WARNING!\n",
      "Some assemblies do not have annotation. \n",
      "Most likely, this is because of an indexing delay. Skipping 3 assemblies.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsYAAAF0CAYAAAAggv9WAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAk8klEQVR4nO3df3CUhYH/8fduAgRICYTwowkoF4HiDyDaoFVU9Nj2biillEPnPA8LopRyHgN6WqxK7SkSpYFKj2KL1ml7dz3Rg9RzmOMmcIXRjGc4yBTtgKKiAgrkpyCEg2S/f/BlNRLMryfsPvB+zThDNpvNe4GYTx6e3Y3E4/E4kiRJ0nkumuwASZIkKRU4jCVJkiQcxpIkSRLgMJYkSZIAh7EkSZIEOIwlSZIkANKTHXDKvn37Ar29nJwcKisrA73NzmBnsOwMThgawc6g2RmsMHSGoRHsDNr53Jmbm3vG93nEWJIkScJhLEmSJAEOY0mSJAlIoXOMPy8ej1NfX09jYyORSKTNH79//36OHTvWCWXB+nxnPB4nGo2SkZHRrvstSZKk9knZYVxfX0+XLl1IT29fYnp6OmlpaQFXBa+5zhMnTlBfX0/37t2TVCVJknT+SdlTKRobG9s9isMuPT2dxsbGZGdIkiSdV1J2GJ/vpxGc7/dfkiTpbEvZYSxJkiSdTaE5V6Hhzkltu34L709b9WL7YwKyatUqvvvd79K1a9dkp0iSJJ33PGKcRE8//TRHjx5NdoYkSZJwGLfo+eefJxaLEYvF+Pu//3s++OADbrrpJmKxGDfffDN79+4FYN68ebz00kuJjxs2bBgAZWVlTJ06lTvvvJPrr7+eu+66i3g8zjPPPMP+/fuZMmUKU6dOTcp9kyRJ0qdCcypFMuzcuZMnn3ySF198kezsbGpqapg3bx433XQTN998M//2b//GQw89xK9+9asvvJ3XX3+djRs3MnDgQL797W9TXl7OzJkz+eUvf8maNWvIyso6S/dIkiRJZ+IR4y/wyiuvMHHiRLKzswHo06cP//u//8t3vvMdAP7qr/6K1157rcXbKSgoIDc3l2g0yqWXXsoHH3zQqd2SJElqO4dxQD773MONjY0cP3488b7PPrguLS2NEydOnPU+SZIkfTFPpfgCY8eOZebMmcyaNStxKkVhYSG///3vmTp1KmvWrOGqq64CYNCgQWzfvp1JkybxX//1X02G8ZlkZmZy+PBhT6WQJEnnjbY809j+Vl4vqGcbC80wbusdTk9P7/CR2a985SvMnTuXqVOnEo1Gueyyy3j00UeZP38+Tz31FNnZ2SxbtgyAW2+9lRkzZhCLxbjxxhvp0aNHi7d/6623csstt9C/f39eeOGFDrVKkiSpYyLxeDye7AiAffv2NXn7yJEjrRqXZxLEMD4bztTZ0fsftJycHCorK5Od0SI7gxOGRrAzaHYGKwydYWgEO4OWzM62vjZFa7TlAGpubu4Z3+c5xpIkSRKtOJWisrKSFStWUFtbSyQSIRaLMWHCBFavXs2GDRvo1asXALfccgtXXHEFAGvXrmXjxo1Eo1FmzJhBQUFBp94JSZIkqaNaHMZpaWlMmzaN/Px8jh49yoIFCxg1ahQA3/zmN5k0qenh8D179lBWVsbSpUupqanhkUce4cknnyQabdvB6RQ5wyNpzvf7L0mSdLa1uFb79OlDfn4+AN27dycvL4/q6uozXr+8vJxrrrmGLl260L9/fwYOHMiuXbvaHhaNhuIc4c5w4sSJNv8gIUmSpI5p07NSHDhwgHfffZehQ4eyY8cO1q9fz+bNm8nPz+e2224jMzOT6urqxMshA2RnZzc7pEtLSyktLQWgqKiInJycJu+Px+NUV1e3exw3NjaG4qhrc51dunRhwIABRCKRJFWdLj09/bQ/o1RkZ3DC0Ah2Bs3OYIWhMwyNYGfQktnZ2qdga4ug7kurh3F9fT3FxcVMnz6dHj168I1vfIOpU6cC8Nxzz/Gb3/yGOXPmtPoTx2IxYrFY4u0zPTIyLS2t1bf5WWF+VGg8HqeqqipJRc0L8+9nKgpDZxgawc6g2RmsMHSGoRHsDFpYOlurLfelw89KceLECYqLi7nuuusSL2jRu3dvotEo0WiU8ePH8/bbbwMnjxB/dtRVV1cnXlJZkiRJSlUtDuN4PM5TTz1FXl4eEydOTFxeU1OT+PVrr73G4MGDASgsLKSsrIzjx49z4MABPvzwQ4YOHdoJ6ZIkSVJwWjyVYufOnWzevJkLLriAe++9Fzj51GyvvPIKu3fvJhKJ0K9fP2bNmgXA4MGDufrqq7n77ruJRqPMnDnTB5JJkiQp5bU4jEeMGMHq1atPu/zUcxY3Z8qUKUyZMqVjZZIkSdJZ5KFcSZIkCYexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAGQ3tIVKisrWbFiBbW1tUQiEWKxGBMmTODw4cMsW7aMgwcP0q9fP+bPn09mZibxeJxnn32Wbdu20a1bN+bMmUN+fv7ZuC+SJElSu7V4xDgtLY1p06axbNkyFi1axPr169mzZw8lJSWMHDmS5cuXM3LkSEpKSgDYtm0bH330EcuXL2fWrFk8/fTTnX0fJEmSpA5rcRj36dMnccS3e/fu5OXlUV1dTXl5OePGjQNg3LhxlJeXA7Blyxauv/56IpEIw4cP55NPPqGmpqYT74IkSZLUcS2eSvFZBw4c4N1332Xo0KHU1dXRp08fAHr37k1dXR0A1dXV5OTkJD6mb9++VFdXJ657SmlpKaWlpQAUFRU1+ZggpKenB36bncHOYNkZnDA0gp1BszNYYegMQyPYGbRkdu7vhNsM6r60ehjX19dTXFzM9OnT6dGjR5P3RSIRIpFImz5xLBYjFosl3q6srGzTx7ckJycn8NvsDHYGy87ghKER7AyancEKQ2cYGsHOoIWls7Xacl9yc3PP+L5WPSvFiRMnKC4u5rrrruOqq64CICsrK3GKRE1NDb169QIgOzu7SVxVVRXZ2dmtjpUkSZKSocVhHI/Heeqpp8jLy2PixImJywsLC9m0aRMAmzZtYsyYMYnLN2/eTDwe580336RHjx6nnUYhSZIkpZoWT6XYuXMnmzdv5oILLuDee+8F4JZbbmHy5MksW7aMjRs3Jp6uDeDyyy9n69atzJ07l65duzJnzpzOvQeSJElSAFocxiNGjGD16tXNvm/hwoWnXRaJRLjjjjs6XiZJkiSdRb7ynSRJkoTDWJIkSQIcxpIkSRLgMJYkSZIAh7EkSZIEOIwlSZIkwGEsSZIkAQ5jSZIkCXAYS5IkSYDDWJIkSQJa8ZLQkiRJCoeGOye16nr7W3l7aatebH9MCHnEWJIkScJhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBEB6S1f4+c9/ztatW8nKyqK4uBiA1atXs2HDBnr16gXALbfcwhVXXAHA2rVr2bhxI9FolBkzZlBQUNB59ZIkSVJAWhzGN9xwA3/5l3/JihUrmlz+zW9+k0mTJjW5bM+ePZSVlbF06VJqamp45JFHePLJJ4lGPTAtSZKk1NbiYr3kkkvIzMxs1Y2Vl5dzzTXX0KVLF/r378/AgQPZtWtXhyMlSZKkztbiEeMzWb9+PZs3byY/P5/bbruNzMxMqqurGTZsWOI62dnZVFdXN/vxpaWllJaWAlBUVEROTk57U5qVnp4e+G12BjuDZWdwwtAIdgbNzmCFoTMMjZD8zv3fuaZ112vDbQ5YW9a+mIA+f2t0xu950I0QXGe7hvE3vvENpk6dCsBzzz3Hb37zG+bMmdOm24jFYsRiscTblZWV7Uk5o5ycnMBvszPYGSw7gxOGRrAzaHYGKwydYWiE8HS2RRjuTxgaoW2dubm5Z3xfu07+7d27N9FolGg0yvjx43n77beBk0eIq6qqEterrq4mOzu7PZ9CkiRJOqvaNYxramoSv37ttdcYPHgwAIWFhZSVlXH8+HEOHDjAhx9+yNChQ4MplSRJkjpRi6dS/PSnP+VPf/oThw4dYvbs2dx888288cYb7N69m0gkQr9+/Zg1axYAgwcP5uqrr+buu+8mGo0yc+ZMn5FCkiRJodDiMJ43b95pl/35n//5Ga8/ZcoUpkyZ0qEoSZIk6WzzcK4kSZKEw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkSAOnJDpAkSeevhjsntep6+9twm2mrXmxfjM57HjGWJEmScBhLkiRJgMNYkiRJAhzGkiRJEuAwliRJkgCHsSRJkgQ4jCVJkiTAYSxJkiQBrXiBj5///Ods3bqVrKwsiouLATh8+DDLli3j4MGD9OvXj/nz55OZmUk8HufZZ59l27ZtdOvWjTlz5pCfn9/pd0KSJEnqqBaPGN9www388Ic/bHJZSUkJI0eOZPny5YwcOZKSkhIAtm3bxkcffcTy5cuZNWsWTz/9dKdES5IkSUFrcRhfcsklZGZmNrmsvLyccePGATBu3DjKy8sB2LJlC9dffz2RSIThw4fzySefUFNT0wnZkiRJUrBaPJWiOXV1dfTp0weA3r17U1dXB0B1dTU5OTmJ6/Xt25fq6urEdT+rtLSU0tJSAIqKipp8XBDS09MDv83OYGew7AxOGBrBzqDZGawwdCa7cX8n3GZn3J/ztTMMjRBcZ7uG8WdFIhEikUibPy4WixGLxRJvV1ZWdjSliZycnMBvszPYGSw7gxOGRrAzaHYGKwydYWhsq7DcnzB0hqER2taZm5t7xve161kpsrKyEqdI1NTU0KtXLwCys7ObhFVVVZGdnd2eTyFJkiSdVe0axoWFhWzatAmATZs2MWbMmMTlmzdvJh6P8+abb9KjR49mT6OQJEmSUk2Lp1L89Kc/5U9/+hOHDh1i9uzZ3HzzzUyePJlly5axcePGxNO1AVx++eVs3bqVuXPn0rVrV+bMmdPpd0CSJEkKQovDeN68ec1evnDhwtMui0Qi3HHHHR2OkiRJks42X/lOkiRJwmEsSZIkAQ5jSZIkCXAYS5IkSUAAL/AhSZJST8Odk1p1vba8ClnaqhfbFyOFhEeMJUmSJBzGkiRJEuAwliRJkgCHsSRJkgQ4jCVJkiTAYSxJkiQBDmNJkiQJcBhLkiRJgMNYkiRJAhzGkiRJEuAwliRJkgCHsSRJkgQ4jCVJkiTAYSxJkiQBDmNJkiQJcBhLkiRJgMNYkiRJAhzGkiRJEuAwliRJkgCHsSRJkgQ4jCVJkiTAYSxJkiQBDmNJkiQJcBhLkiRJgMNYkiRJAhzGkiRJEuAwliRJkgCHsSRJkgQ4jCVJkiTAYSxJkiQBDmNJkiQJcBhLkiRJgMNYkiRJAhzGkiRJEuAwliRJkgCHsSRJkgQ4jCVJkiTAYSxJkiQBDmNJkiQJcBhLkiRJAKQnO0CSJICGOye1+rr7W3m9tFUvti9G0nnJI8aSJEkSDmNJkiQJcBhLkiRJgMNYkiRJAnzwnSRJbeKDBKVzl0eMJUmSJDp4xPjv/u7vyMjIIBqNkpaWRlFREYcPH2bZsmUcPHiQfv36MX/+fDIzM4PqlSRJkjpFh0+l+NGPfkSvXr0Sb5eUlDBy5EgmT55MSUkJJSUl/O3f/m1HP40kSZLUqQI/laK8vJxx48YBMG7cOMrLy4P+FJIkSVLgOnzEeNGiRQB8/etfJxaLUVdXR58+fQDo3bs3dXV1zX5caWkppaWlABQVFZGTk9PRlCbS09MDv83OYGew7AxOGBrBzqAls7O1D1Rri864L3YGy85gBd0ZhkYIrrNDw/iRRx4hOzuburo6Hn30UXJzc5u8PxKJEIlEmv3YWCxGLBZLvF1ZWdmRlNPk5OQEfpudwc5g2RmcMDSCnUELS2drheW+2BksO4MThkZoW+fn9+pndehUiuzsbACysrIYM2YMu3btIisri5qaGgBqamqanH8sSZIkpap2D+P6+nqOHj2a+PUf//hHLrjgAgoLC9m0aRMAmzZtYsyYMcGUSpIkSZ2o3adS1NXV8ZOf/ASAhoYGrr32WgoKCrjoootYtmwZGzduTDxdmyRJkpTq2j2MBwwYwJIlS067/Etf+hILFy7sUJQkSZJ0tvnKd5IkSRIOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkS4DCWJEmSAIexJEmSBDiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCHMaSJEkSAOnJDpCkMGu4c1Krrre/DbeZturF9sVIkjrEI8aSJEkSDmNJkiQJcBhLkiRJgMNYkiRJAnzwnRSY1j4IC1r/QCwfhCVJ0tnjEWNJkiQJjxhL5x2fXkySpOZ5xFiSJEnCYSxJkiQBDmNJkiQJcBhLkiRJgA++Uwj4NGhSx/g1JEmt4xFjSZIkCYexJEmSBHgqhaQU5fMtS5LONo8YS5IkSTiMJUmSJMBhLEmSJAEOY0mSJAlwGEuSJEmAw1iSJEkCOvHp2ioqKnj22WdpbGxk/PjxTJ48ubM+VUoK+qmmfJopSZKkztUpw7ixsZFnnnmGBx98kL59+3L//fdTWFjIoEGDOuPTqZ18mVhJkqRPdcow3rVrFwMHDmTAgAEAXHPNNZSXlwcyjH3Sf0mSJHWGSDwejwd9o6+++ioVFRXMnj0bgM2bN/PWW28xc+bMxHVKS0spLS0FoKioKOgESZIkqU2S9uC7WCxGUVFRp43iBQsWdMrtBs3OYNkZnDA0gp1BszNYYegMQyPYGTQ7m9cpwzg7O5uqqqrE21VVVWRnZ3fGp5IkSZIC0SnD+KKLLuLDDz/kwIEDnDhxgrKyMgoLCzvjU0mSJEmB6JQH36WlpXH77bezaNEiGhsbufHGGxk8eHBnfKozisViZ/XztZedwbIzOGFoBDuDZmewwtAZhkawM2h2Nq9THnwnSZIkhY2vfCdJkiThMJYkSZIAh7EkSZIEOIwlSZIk4Bwexv/6r/+a7ARJkiSFSKc8XdvZ9qtf/eq0yzZv3kx9fT0At99++9lOalZlZSW9evWia9euxONx/vCHP/Duu+8yaNAgxo8fT1paWrITAdiyZQujRo2ia9euyU75QvX19VRUVFBZWUk0GiU3N5dRo0YRjabWz3t79+6lvLyc6upq4OQL4BQWFjJo0KAkl7XOf//3f3PjjTcmOyNh7969VFdXM2zYMDIyMhKXV1RUUFBQkLywz9m1axcAQ4cOZc+ePVRUVJCbm8sVV1yR5LIv9k//9E/cddddyc74Qjt27GDXrl0MHjyY0aNHJzsHgLfeeou8vDx69OjB//3f/1FSUsI777zDoEGDmDJlCj169Eh2IgDr1q3jyiuvJCcnJ9kpX+jEiRO88sor9OnTh1GjRvHyyy+zc+dO8vLyiMVipKenznzZv38///M//0NVVRXRaJQvf/nLXHvttSnzZ36K34ta55x4urbvf//7XHzxxYwePZpTd+e3v/0t06ZNA+CGG25IYt2n7rnnHh577DG6devGP//zP7N//37GjBnD66+/DsCcOXOSXHjSrbfeSkZGBgUFBYwdO5aCgoKUG5tlZWX8x3/8BxdeeCFvvPEGw4cPJx6P8/777zN37lwuuOCCZCcCUFJSwiuvvMLYsWMTr/5YXV2duGzy5MnJDWyF73//+6xcuTLZGcDJb+rr168nLy+P9957j+nTpzNmzBgAfvCDH/D4448nufCk559/noqKChoaGhg1ahRvvfUWl156Kdu3b2f06NFMmTIl2YkAp/1+xeNx3njjDS677DLg5O9pKrj//vtZvHgxAKWlpaxfv54rr7ySP/7xj3z1q19Nia+ju+++myVLlpCWlsYvfvELunXrxte+9jW2b9/Oe++9xz/8wz8kOxGA7373u2RkZDBgwADGjh3L1VdfTa9evZKddZrly5fT0NDAsWPH6NmzJ/X19Vx11VVs376deDyeMj+8rVu3jq1bt3LxxRezbds2hgwZQs+ePXnttde44447uPTSS5OdCPi9qC1S50euDli6dCnPPfccFRUVTJs2jezsbF544YWUGcSnNDY20q1bNwC2b9/O4sWLiUajXH/99dx7771JrvtUXl4eCxcu5NVXX+Wll15i5cqVjBkzhmuvvZZLLrkk2XkArFmzhkWLFtGtWzc+/vhjfvazn/HAAw/w3nvv8ctf/pJHH3002YnAyZ9wi4uLTzu6MXHiRO6+++6U+Z/Rmb5px+Nx6urqznLNmW3YsIHHH3+cjIwMDhw4wNKlSzl48CATJkwglX7Gf/XVV1myZAnHjx9n1qxZrFy5kh49ejBp0iR++MMfpswwrq6uJi8vj/HjxxOJRIjH47zzzjt861vfSnZaEw0NDYlfb9iwgYceeohevXrxrW99iwceeCAlvo7i8XjiX/3eeeedxA8dI0aMSKn/vw8YMICioiK2b99OWVkZq1evJj8/n7Fjx3LVVVfRvXv3ZCcC8P777/OTn/yEhoYGZs+ezS9+8Qui0SjXXXddSv1+btiwgSVLlhCNRpk4cSKLFy/m4Ycf5utf/zpPPPEETzzxRLITAb8XtcU5MYy7d+/O9OnTeeedd/jZz37G5ZdfnlLfJE/Jycnh9ddf57LLLqNfv35UVVXRr18/Dh06lOy0JiKRCJmZmcRiMWKxGLW1tZSVlfEv//IvVFdXp8TRw3g8njjVIyMjI/EFc+GFF3L06NFkpjURiUSoqamhX79+TS6vqakhEokkqep0dXV1PPDAA/Ts2bPJ5fF4nIceeihJVaeLx+OJ0yf69+/Pww8/THFxMQcPHkypr/m0tDSi0SjdunVjwIABiX9S7dq1a0r9uS9evJh169axZs0apk2bxpAhQ+jatWvK/AB8Sjwe5/Dhw8TjceLxeOIIZ0ZGRsqcgjZ48ODEP/VeeOGFvP3221x00UXs27cvpf7ZPxKJEI1GGT16NKNHj+bEiRNUVFTw8ssv89vf/pZnnnkm2YnAyT/zEydOUF9fz7Fjxzhy5AiZmZkcP368yQ9KqaChoYFoNMrx48cTp3Dm5OSkVKffi1ovdb5aA5Cfn8/ChQtZv349X/nKV5Kdc5rvfe97rFixgueff57u3btz3333MWTIED755BNuu+22ZOclfH5g9O7dmwkTJjBhwgQOHjyYpKqmLr/8ch577DEuvvhiKioq+NrXvgaQ+OaZKqZPn84//uM/8uUvf5m+ffsCJ881/+ijj5g5c2aS6z51xRVXUF9fz5AhQ057XyqNpKysLHbv3p3ozMjIYMGCBaxcuZL3338/uXGfkZ6ezrFjx+jWrRtFRUWJy48cOZJSpyWdOsp19dVX8+tf/5qsrKyU+mZ+ypEjR1iwYAHxeDzxDb5Pnz7U19enzNf77NmzefbZZ1mzZg1f+tKXePDBB+nbty99+/ble9/7XrLzEj7/+5Wenk5hYSGFhYUcO3YsSVWnu/HGG5k3bx6NjY389V//NUuXLqV///689dZbXHPNNcnOSxg/fjz3338/Q4cOZceOHXz7298G4OOPPyYzMzPJdZ/ye1HrnRPnGJ9SW1vb5KTy3r17JzfoDPbs2cOHH35IQ0MDffv25aKLLkqpb5ZvvPFGypwX9UW2bt3Knj17GDJkCKNGjQJOnq7S0NBAly5dklz3qcbGRnbt2tXk7+bQoUNT6s88LKqqqkhLS2v2a3vHjh2MGDHi7Ec14/jx483+Hfz444+pra1NmXPgP2/r1q3s2LGDv/mbv0l2SqscO3aMuro6+vfvn+yUhCNHjnDgwAEaGxtT8vvQvn37yM3NTXZGq3z2/5mffPIJ27dvJycnh6FDhya5rKkPPviAvXv3MnjwYPLy8pKdc0Z+L2qdc2IY7969m1WrVnHkyJHESeVVVVX07NmTmTNnkp+fn+TCpsIy4MPQGYbGM6mvr2/yrAqpys5g2RmsMHSGoRHsDFpYOtXUOTGM7733XmbNmsWwYcOaXP7mm2+yatUqlixZkqSypsIy4MPQGYbGlqTSsz18ETuDZWewwtAZhkawM2ip1HnqgenV1dUUFBRw6623Jk71+OyzviRbKnSeE+cYHzt27LRRDDB8+PDEifCpYMWKFWcc8CtXrkyZAR+GzjA0Arz00kvNXh6Px1Pq76adwbIzWGHoDEMj2Bm0sHQ+/fTT3HTTTQwbNowNGzawcOFC7rvvPgYOHJhSjytIhc5z4sSSgoICFi9eTFlZGTt37mTnzp2UlZWxePHilHrC/7AM+DB0hqER4He/+x2HDx/m6NGjTf5LpQcNgZ1BszNYYegMQyPYGbSwdNbX11NQUEDPnj2ZNGkSt99+O4899hhvvvlmSj0rRSp0nhNHjG+//Xa2bdt22iu6/MVf/EVKvcrUqQE/bty4xKNCq6qq2LRpU0oN+DB0hqER4M/+7M+48sormz21Y+PGjUkoap6dwbIzWGHoDEMj2Bm0sHTCyQeGnnrqyMsuu4x77rmH4uJiDh8+nOSyppLdeU6cYxwmzQ34wsLClBrwEI7OMDTu27ePzMzMZl9Zqra2NmUeLGhnsOwMVhg6w9AIdgYtLJ0vv/wy/fv3Z/jw4U0ur6ys5IUXXmD27NlJKmsqFTrPiWF85MgR1q5dy5YtW6itrSUSiZCVlUVhYSGTJ08+7YmiJUmSpM87J4bxokWLuPTSS7nhhhsSP53V1tbyhz/8gddff50HH3wwuYH/X1gGfBg6w9AIn3aWl5dTV1dnZwfZGSw7z69GsDNodgYrJTrj54C5c+e2631n26OPPhpfu3ZtvKamJnFZTU1NfO3atfFHHnkkeWGfE4bOMDTG43YGzc5g2RmcMDTG43YGzc5gpULnOfGsFP369eP3v/89tbW1ictqa2spKSkhJycneWGfc+DAASZPntzknKPevXszefLklHmpZQhHZxgawc6g2RksO4MThkawM2h2BisVOs+JYTxv3jwOHTrEww8/zIwZM5gxYwY//vGPOXz4MPPnz092XkJYBnwYOsPQCHYGzc5g2RmcMDSCnUGzM1ip0HlOnGMMsHfvXqqqqhg+fHiTl2CsqKhImafvOnz4MCUlJWzZsoW6ujrg5E9CX/3qV5k8eXLi1V2SLQydYWgEO4NmZ7DsDE4YGsHOoNkZrFToPCeG8bp161i/fj15eXm89957TJ8+nTFjxgDwgx/8gMcffzzJhZ8Kw4CHcHSGoRHsDJqdwbIzOGFoBDuDZmewkt15TpxKsWHDBh5//HHuu+8+fvSjH/Hv//7vrFu3DiClXnlm3bp1PPHEE/znf/4n99xzD+Xl5Yn3/e53v0tiWVNh6AxDI9gZNDuDZWdwwtAIdgbNzmClQuc58cp38Xg88VNF//79efjhhykuLubgwYMpNYxPDfiMjAwOHDjA0qVLOXjwIBMmTLCzjcLQCHYGzc5g2RmcMDSCnUGzM1ip0HlOHDHOyspi9+7dibczMjJYsGABhw4d4v33309e2Oc0N+C3bdvGr3/965T6ixmGzjA0gp1BszNYdgYnDI1gZ9DsDFYqdJ4Tw/iuu+467WUX09LSuOuuu/jxj3+cnKhmhGXAh6EzDI1gZ9DsDJadwQlDI9gZNDuDlQqd58SD78KiqqqKtLS0Zl87fceOHYwYMeLsRzUjDJ1haAQ7g2ZnsOwMThgawc6g2RmsVOh0GEuSJEmcI6dSSJIkSR3lMJYkSZJwGEuSJEmAw1iSJEkC4P8BT+CbszD1oK8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 864x432 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def annotation_year(annotation_metadata):\n",
    "    return datetime.strptime(annotation_metadata.release_date, '%Y-%M-%d').year\n",
    "\n",
    "annots_by_year = Counter()\n",
    "no_annot_assms = []\n",
    "for d in map(lambda d: d.assembly, genome_summary.assemblies):\n",
    "    if not d.annotation_metadata:\n",
    "        no_annot_assms.append(d.assembly_accession)\n",
    "        continue\n",
    "    annots_by_year[annotation_year(d.annotation_metadata)] += 1\n",
    "\n",
    "\n",
    "if len(no_annot_assms) > 0:\n",
    "    print(dedent(f'''\n",
    "    WARNING!\n",
    "    Some assemblies do not have annotation. \n",
    "    Most likely, this is because of an indexing delay. Skipping {len(no_annot_assms)} assemblies.\n",
    "'''))\n",
    "\n",
    "df = pd.DataFrame.from_dict(annots_by_year, orient='index', columns=['count']).sort_index()\n",
    "df.plot(kind='bar', y='count', figsize=(12,6))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can make a list of the genome assemblies that were annotated in 2020 and count the number of genomes in the list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "187 genome assemblies were annotated in 2020.\n",
      "Here are the first five genome assemblies in the list: ['GCF_012848655.1', 'GCF_014648715.1', 'GCF_013342945.1', 'GCF_013867555.1', 'GCF_013867605.1']\n"
     ]
    }
   ],
   "source": [
    "year_of_interest = 2020\n",
    "genome_accessions = []\n",
    "for d in map(lambda d: d.assembly, genome_summary.assemblies):\n",
    "    if d.annotation_metadata:\n",
    "        if annotation_year(d.annotation_metadata) == year_of_interest:\n",
    "            genome_accessions.append(d.assembly_accession)\n",
    "        \n",
    "print(f'{len(genome_accessions)} genome assemblies were annotated in {year_of_interest}.')\n",
    "print(f\"Here are the first five genome assemblies in the list: {genome_accessions[0:5]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We could download genome datasets for all of the returned assemblies, based on the accessions stored in `genome_accessions`.  If there are quite a few of them, a `download_api.download_assembly_package_post()` request would be required (see docs here)[https://github.com/ncbi/datasets/blob/master/client_docs/python/docs/DownloadApi.md#download_assembly_package_post]. \n",
    "\n",
    "For demonstration purposes, we'll just download genome datasets for five genome assemblies through a GET request."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Download genome datasets for 5 accessions: ['GCF_013778545.1', 'GCF_014654865.1', 'GCF_014982905.1', 'GCF_015669875.1', 'GCF_015708055.1']\n",
      "Download saved to ncbi_genomes.zip\n",
      "Archive:  ncbi_genomes.zip\n",
      "  Length      Date    Time    Name\n",
      "---------  ---------- -----   ----\n",
      "      661  07-16-2021 15:02   README.md\n",
      "     6167  07-16-2021 15:02   ncbi_dataset/data/assembly_data_report.jsonl\n",
      "  2323408  07-16-2021 15:02   ncbi_dataset/data/GCF_013778545.1/GCF_013778545.1_ASM1377854v1_genomic.fna\n",
      "  1242442  07-16-2021 15:02   ncbi_dataset/data/GCF_013778545.1/genomic.gff\n",
      "   765683  07-16-2021 15:02   ncbi_dataset/data/GCF_013778545.1/protein.faa\n",
      "  2094903  07-16-2021 15:02   ncbi_dataset/data/GCF_014654865.1/GCF_014654865.1_Biofab_BC5_1.1_genomic.fna\n",
      "  1129307  07-16-2021 15:02   ncbi_dataset/data/GCF_014654865.1/genomic.gff\n",
      "   732940  07-16-2021 15:02   ncbi_dataset/data/GCF_014654865.1/protein.faa\n",
      "  1965961  07-16-2021 15:02   ncbi_dataset/data/GCF_014982905.1/GCF_014982905.1_ASM1498290v1_genomic.fna\n",
      "  1048475  07-16-2021 15:02   ncbi_dataset/data/GCF_014982905.1/genomic.gff\n",
      "   693726  07-16-2021 15:02   ncbi_dataset/data/GCF_014982905.1/protein.faa\n",
      "  2312795  07-16-2021 15:02   ncbi_dataset/data/GCF_015669875.1/GCF_015669875.1_ASM1566987v1_genomic.fna\n",
      "  1177206  07-16-2021 15:02   ncbi_dataset/data/GCF_015669875.1/genomic.gff\n",
      "   812033  07-16-2021 15:02   ncbi_dataset/data/GCF_015669875.1/protein.faa\n",
      "  2422623  07-16-2021 15:02   ncbi_dataset/data/GCF_015708055.1/GCF_015708055.1_ASM1570805v1_genomic.fna\n",
      "  1340180  07-16-2021 15:02   ncbi_dataset/data/GCF_015708055.1/genomic.gff\n",
      "   804635  07-16-2021 15:02   ncbi_dataset/data/GCF_015708055.1/protein.faa\n",
      "    11553  07-16-2021 15:02   ncbi_dataset/data/GCF_013778545.1/sequence_report.jsonl\n",
      "     9822  07-16-2021 15:02   ncbi_dataset/data/GCF_014654865.1/sequence_report.jsonl\n",
      "     9527  07-16-2021 15:02   ncbi_dataset/data/GCF_014982905.1/sequence_report.jsonl\n",
      "     9335  07-16-2021 15:02   ncbi_dataset/data/GCF_015669875.1/sequence_report.jsonl\n",
      "    16003  07-16-2021 15:02   ncbi_dataset/data/GCF_015708055.1/sequence_report.jsonl\n",
      "     2539  07-16-2021 15:02   ncbi_dataset/data/dataset_catalog.json\n",
      "---------                     -------\n",
      " 20931924                     23 files\n",
      "CPU times: user 47.1 ms, sys: 44.1 ms, total: 91.3 ms\n",
      "Wall time: 1.46 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "## Download data\n",
    "# Let's use accessions #21-25 in the list because these all have at least one crispr gene\n",
    "print(f'Download genome datasets for 5 accessions: {genome_accessions[20:25]}')\n",
    "dl_response = genome_api.download_assembly_package(\n",
    "    accessions=genome_accessions[20:25],\n",
    "    exclude_sequence=False,\n",
    "    include_annotation_type=['GENOME_GFF', 'PROT_FASTA'],\n",
    "    _preload_content=False )\n",
    "\n",
    "## write to a zip file \n",
    "zipfn = 'ncbi_genomes.zip'\n",
    "with open(zipfn, 'wb') as f:\n",
    "    f.write(dl_response.data)\n",
    "print(f'Download saved to {zipfn}')\n",
    "!unzip -l {zipfn}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we're going to look for crispr genes in these genome datasets, and extract the corresponding crispr gene and protein sequences.  \n",
    "We're going to break this down into a few steps.  \n",
    "First, let's make a list of the data files that are available for each genome assembly. We can use the `dataset_catalog.json` file to tell us which data files are available for each genome assembly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Catalog found with metadata for 5 assemblies\n",
      "Assemblies:\n",
      "GCF_013778545.1, GCF_014654865.1, GCF_014982905.1, GCF_015669875.1, GCF_015708055.1\n",
      "Files for type: GFF3\n",
      "defaultdict(<class 'list'>,\n",
      "            {   'GCF_013778545.1': [   'ncbi_dataset/data/GCF_013778545.1/genomic.gff'],\n",
      "                'GCF_014654865.1': [   'ncbi_dataset/data/GCF_014654865.1/genomic.gff'],\n",
      "                'GCF_014982905.1': [   'ncbi_dataset/data/GCF_014982905.1/genomic.gff'],\n",
      "                'GCF_015669875.1': [   'ncbi_dataset/data/GCF_015669875.1/genomic.gff'],\n",
      "                'GCF_015708055.1': [   'ncbi_dataset/data/GCF_015708055.1/genomic.gff']})\n",
      "Files for type: FASTA\n",
      "defaultdict(<class 'list'>, {})\n",
      "Files for type: GENOMIC_NUCLEOTIDE_FASTA\n",
      "defaultdict(<class 'list'>,\n",
      "            {   'GCF_013778545.1': [   'ncbi_dataset/data/GCF_013778545.1/GCF_013778545.1_ASM1377854v1_genomic.fna'],\n",
      "                'GCF_014654865.1': [   'ncbi_dataset/data/GCF_014654865.1/GCF_014654865.1_Biofab_BC5_1.1_genomic.fna'],\n",
      "                'GCF_014982905.1': [   'ncbi_dataset/data/GCF_014982905.1/GCF_014982905.1_ASM1498290v1_genomic.fna'],\n",
      "                'GCF_015669875.1': [   'ncbi_dataset/data/GCF_015669875.1/GCF_015669875.1_ASM1566987v1_genomic.fna'],\n",
      "                'GCF_015708055.1': [   'ncbi_dataset/data/GCF_015708055.1/GCF_015708055.1_ASM1570805v1_genomic.fna']})\n",
      "Files for type: PROTEIN_FASTA\n",
      "defaultdict(<class 'list'>,\n",
      "            {   'GCF_013778545.1': [   'ncbi_dataset/data/GCF_013778545.1/protein.faa'],\n",
      "                'GCF_014654865.1': [   'ncbi_dataset/data/GCF_014654865.1/protein.faa'],\n",
      "                'GCF_014982905.1': [   'ncbi_dataset/data/GCF_014982905.1/protein.faa'],\n",
      "                'GCF_015669875.1': [   'ncbi_dataset/data/GCF_015669875.1/protein.faa'],\n",
      "                'GCF_015708055.1': [   'ncbi_dataset/data/GCF_015708055.1/protein.faa']})\n"
     ]
    }
   ],
   "source": [
    "## function to make list of files \n",
    "from ncbi.datasets.v1 import catalog_pb2\n",
    "import pprint\n",
    "pp = pprint.PrettyPrinter(indent=4)\n",
    "\n",
    "def retrieve_data_catalog(zip_file):\n",
    "    with zipfile.ZipFile(zip_file, 'r') as zip:\n",
    "        data_catalog = json.loads(zip.read('ncbi_dataset/data/dataset_catalog.json'))\n",
    "    print(f\"Catalog found with metadata for {len(data_catalog['assemblies'])-1} assemblies\")\n",
    "    return data_catalog\n",
    "\n",
    "def get_assemblies(data_catalog):\n",
    "    return [x['accession'] for x in data_catalog['assemblies'] if 'accession' in x]\n",
    "\n",
    "# Temporary hack to support GENOMIC_NUCLEOTIDE_FASTA & PROTEIN_FASTA \n",
    "# which will be present in the next release\n",
    "def get_file_list(data_catalog, desired_filetype):\n",
    "    desired_filetype = desired_filetype.upper()\n",
    "    if desired_filetype not in catalog_pb2.File.FileType.keys():\n",
    "        raise Exception(f'Filetype {desired_filetype} is invalid.')\n",
    "    \n",
    "    files = defaultdict(list)\n",
    "    for asm in data_catalog['assemblies']:\n",
    "        if 'accession' in asm:\n",
    "            acc = asm['accession'] \n",
    "            for f in asm['files']:\n",
    "                filepath = os.path.join('ncbi_dataset', 'data', f['filePath'])\n",
    "                if f['fileType'] == 'FASTA' and desired_filetype in ('GENOMIC_NUCLEOTIDE_FASTA') and filepath.endswith('fna'):\n",
    "                    files[acc].append(filepath)\n",
    "                    continue\n",
    "                if f['fileType'] == 'GFF3':\n",
    "                    if desired_filetype in ('PROTEIN_FASTA') and filepath.endswith('faa'):\n",
    "                        files[acc].append(filepath)\n",
    "                    if desired_filetype in ('GFF3') and filepath.endswith('gff'):\n",
    "                        files[acc].append(filepath)\n",
    "                    continue\n",
    "                if f['fileType'] == desired_filetype:\n",
    "                    files[acc].append(filepath)\n",
    "        \n",
    "    return files\n",
    "\n",
    "data_catalog = retrieve_data_catalog(zipfn)\n",
    "print(f'Assemblies:')\n",
    "print(', '.join(get_assemblies(data_catalog)))\n",
    "for file_type in ['GFF3', 'FASTA', 'GENOMIC_NUCLEOTIDE_FASTA', 'PROTEIN_FASTA']:\n",
    "    file_list = get_file_list(data_catalog, file_type)\n",
    "    print(f'Files for type: {file_type}')\n",
    "    pp.pprint(file_list)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We're going to create temporary files to store the genome annotation (gff3 format) and protein sequence data. We will put the files in a temporary directory named tempfiles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## setting up files and directories\n",
    "\n",
    "## temporary files; will be deleted at the end\n",
    "temp_dir = 'tempfiles'\n",
    "temp_gff = temp_dir + '/temp.gff'\n",
    "temp_fa = temp_dir + '/temp.fa'\n",
    "\n",
    "!mkdir -p {temp_dir}\n",
    "\n",
    "## final output files \n",
    "genes_fn = 'crispr_genes.fna'\n",
    "prots_fn = 'crispr_proteins.faa'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we need a few functions to process the data. One of these functions will create a gff3 database in memory using the gffutils python package. Another combines fasta files, a third function will extract gene information by gene symbol from the gff3 database. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_gff3_db(files_by_assembly, temp_file, zfh):\n",
    "    '''\n",
    "    create gff3 db in memory, per assembly\n",
    "    okay for bacterial assemblies but use caution\n",
    "    when parsing large assemblies like human\n",
    "    '''\n",
    "\n",
    "    db = {}\n",
    "    for assembly_accession, files in files_by_assembly.items():\n",
    "        with open(temp_file, 'wb') as f:\n",
    "            for file in files:\n",
    "                print(f'\\tWrite {file} to {temp_file}')\n",
    "                f.write(zfh.read(file))\n",
    "            db[assembly_accession] = gffutils.create_db(\n",
    "                temp_file,\n",
    "                dbfn = ':memory:',\n",
    "                force=True,\n",
    "                keep_order=True,\n",
    "                merge_strategy='merge',\n",
    "                sort_attribute_values=True)\n",
    "\n",
    "\n",
    "    return db\n",
    "\n",
    "\n",
    "def combine_fasta(files_by_assembly, temp_file, zfh):\n",
    "    '''\n",
    "    Combine fasta for *all* FASTA files, nt & protein\n",
    "    '''\n",
    "    with open(temp_file, 'wb') as f:\n",
    "        for assembly, files in files_by_assembly.items():\n",
    "            for file in files:\n",
    "                print(f'\\tAppending {file} to {temp_file}')\n",
    "                f.write(zfh.read(file))\n",
    "\n",
    "    print(f'Create FASTA object for {temp_file}.')\n",
    "    return Fasta(temp_file, read_long_names=False, duplicate_action='first')\n",
    "    \n",
    "\n",
    "def extract_genes(gff3_db, assemblies, desired_genes):\n",
    "    crispr_order = defaultdict(list)\n",
    "    crispr_genes = {}\n",
    "\n",
    "    for assembly_accession in assemblies:\n",
    "        if assembly_accession in assemblies:\n",
    "            for gene in gff3_db[assembly_accession].features_of_type('gene'):\n",
    "                gene_name = gene.attributes.get('Name', None)[0]\n",
    "                if gene_name[:4] not in desired_genes:\n",
    "                    continue\n",
    "                gene_range = (gene.start, gene.end)\n",
    "                prot_acc = None\n",
    "                if gene.attributes['gene_biotype'][0] == 'protein_coding':\n",
    "                    cds = list(gff3_db[assembly_accession].children(gene, featuretype='CDS'))\n",
    "                    prot_acc = cds[0].attributes.get('protein_id', None)[0]\n",
    "\n",
    "                crispr_genes[gene_name] = ([gene.chrom, gene.strand, gene_range, prot_acc])\n",
    "                crispr_order[assembly_accession].append(gene_name)\n",
    "    return crispr_genes, crispr_order\n",
    "\n",
    "def write_fasta(fh, defline, content):\n",
    "    fh.write('>' + defline + '\\n')\n",
    "    fh.write(content + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's use the above functions to process the datasets zip file. First, create the gff3 database, then extract crispr gene information from the gff3 data, then for each gene, store the crispr gene data and write the fasta sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Create GFF3 Database\n",
      "\tWrite ncbi_dataset/data/GCF_013778545.1/genomic.gff to tempfiles/temp.gff\n",
      "\tWrite ncbi_dataset/data/GCF_014654865.1/genomic.gff to tempfiles/temp.gff\n",
      "\tWrite ncbi_dataset/data/GCF_014982905.1/genomic.gff to tempfiles/temp.gff\n",
      "\tWrite ncbi_dataset/data/GCF_015669875.1/genomic.gff to tempfiles/temp.gff\n",
      "\tWrite ncbi_dataset/data/GCF_015708055.1/genomic.gff to tempfiles/temp.gff\n",
      "Extract crispr genes (12 genes)\n",
      "Write genes to crispr_genes.fna\n",
      "Create genomes object\n",
      "\tAppending ncbi_dataset/data/GCF_013778545.1/GCF_013778545.1_ASM1377854v1_genomic.fna to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_014654865.1/GCF_014654865.1_Biofab_BC5_1.1_genomic.fna to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_014982905.1/GCF_014982905.1_ASM1498290v1_genomic.fna to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_015669875.1/GCF_015669875.1_ASM1566987v1_genomic.fna to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_015708055.1/GCF_015708055.1_ASM1570805v1_genomic.fna to tempfiles/temp.fa\n",
      "Create FASTA object for tempfiles/temp.fa.\n",
      "Create proteome object\n",
      "\tAppending ncbi_dataset/data/GCF_013778545.1/protein.faa to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_014654865.1/protein.faa to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_014982905.1/protein.faa to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_015669875.1/protein.faa to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_015708055.1/protein.faa to tempfiles/temp.fa\n",
      "Create FASTA object for tempfiles/temp.fa.\n",
      "CPU times: user 5.35 s, sys: 64.2 ms, total: 5.41 s\n",
      "Wall time: 5.47 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "crispr_genes = set(['cas3', 'cse1', 'cse2', 'cas7', 'cas5', 'cas6', 'cas4', 'cas1', 'cas2', 'cas5', 'cas8', 'cas9', 'csn2'])\n",
    "\n",
    "\n",
    "## create empty files to add data to\n",
    "open(genes_fn, 'w').close()\n",
    "open(prots_fn, 'w').close()\n",
    "\n",
    "with zipfile.ZipFile(zipfn, 'r') as zfh:\n",
    "\n",
    "    print('Create GFF3 Database')\n",
    "    gff3_db = create_gff3_db(get_file_list(data_catalog, 'GFF3'), temp_gff, zfh)\n",
    "    \n",
    "    print(f'Extract crispr genes ({len(crispr_genes)} genes)')\n",
    "    crispr_genes, crispr_order = extract_genes(gff3_db, get_assemblies(data_catalog), crispr_genes)\n",
    "    print(f'Write genes to {genes_fn}')\n",
    "\n",
    "    print('Create genomes object')\n",
    "    genomes = combine_fasta(get_file_list(data_catalog, 'GENOMIC_NUCLEOTIDE_FASTA'), temp_fa, zfh)\n",
    "    \n",
    "    # Write all genes to genes_fn\n",
    "    with open(genes_fn, 'w') as fh:\n",
    "        for gene_name, gene_info in crispr_genes.items():\n",
    "            chrom, strand, gene_range, prot_acc = gene_info\n",
    "            reverse_complement = True if strand == '-' else False\n",
    "            gene_fasta = genomes.get_seq(chrom, gene_range[0], gene_range[1], rc=reverse_complement)\n",
    "            write_fasta(fh, f'{gene_fasta.fancy_name}|{gene_name}', gene_fasta.seq)\n",
    "\n",
    "    print('Create proteome object')\n",
    "    proteome = combine_fasta(get_file_list(data_catalog, 'PROTEIN_FASTA'), temp_fa, zfh)\n",
    "    \n",
    "    # Write protein FASTA for each CRISPR gene\n",
    "    with open(prots_fn, 'w') as fh:\n",
    "        for gene_name, gene_info in crispr_genes.items():\n",
    "            chrom, strand, gene_range, prot_acc = gene_info\n",
    "            if prot_acc is not None:\n",
    "                prot_fasta = proteome[prot_acc][:]\n",
    "                write_fasta(fh, f'{prot_fasta.name}|{gene_name}', prot_fasta.seq)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "crispr_genes.fna and crispr_proteins.faa contain the genomic and protein FASTA sequences, respectively, for the crispr genes we identified in all of the five genome assemblies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_015669875.1 ['cas3', 'cas7e', 'cas5e', 'cas6e', 'cas1e', 'cas2e']\n",
      "GCF_015708055.1 ['csn2', 'csn2', 'cas2', 'cas1']\n"
     ]
    }
   ],
   "source": [
    "## order of crispr genes in various assemblies\n",
    "for k,v in crispr_order.items():\n",
    "    print(k, v)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  },
  "metadata": {
   "interpreter": {
    "hash": "b3ce2b738dea48d83b79aea2ce81e31fdf4bb1e268147ae937b9a62885b78b59"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
