{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using `ncbi.datasets` library to download annotation and parse GFF3 files \n",
    "\n",
    "Here, we will use `ncbi.datasets` to download annotation data for some assemblies and parse them to extract specific information.  \n",
    "\n",
    "Specifically, \n",
    "1. query NCBI for _Lactobacillus_ assemblies and download the assembly descriptors for all available assemblies\n",
    "2. parse assembly descriptors to make a list of assemblies annotated in 2020\n",
    "3. download genome sequence and annotation (gff3 and protein fasta) for those assemblies\n",
    "4. parse gff3 files to extract information about crispr genes\n",
    "  1. use that information to extract gene sequences from genome fasta files \n",
    "  2. use that information to extract protein sequences from protein fasta files "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "After importing the various python modules we will need, create the assembly and download API instances."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict, Counter\n",
    "from datetime import datetime\n",
    "import json\n",
    "import os\n",
    "from textwrap import dedent\n",
    "import zipfile\n",
    "\n",
    "import gffutils\n",
    "import matplotlib.pyplot as plt\n",
    "import ncbi.datasets\n",
    "import pandas as pd\n",
    "from pyfaidx import Fasta\n",
    "\n",
    "plt.style.use('ggplot')\n",
    "\n",
    "genome_api = ncbi.datasets.GenomeApi(ncbi.datasets.ApiClient())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Group assemblies based on assembly annotation meta-data\n",
    "Let's look at RefSeq *lactobacillus* assemblies, and see how we can use the annotation meta-data to organize the assembly data.\n",
    "\n",
    "Using the `ncbi.datasets` library, we first check the number of available RefSeq *lactobacillus* assemblies.  When we use `limit='none'`, no descriptors are returned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'descriptors' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-4309d0162f01>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m     filters_refseq_only=True)\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Number of assemblies: {descriptors.total_count}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'descriptors' is not defined"
     ]
    }
   ],
   "source": [
    "taxid = 1578 ## lactobacillus\n",
    "assemblies_and_messages = genome_api.assembly_descriptors_by_taxon(\n",
    "    taxon=taxid,\n",
    "    limit='none',\n",
    "    filters_refseq_only=True)\n",
    "\n",
    "print(f\"Number of assemblies: {descriptors.total_count}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we know a reasonable number of assemblies are available, let's fetch all the data.  Be patient, this can take up to half a minute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.85 s, sys: 12.8 ms, total: 1.87 s\n",
      "Wall time: 2.96 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "## download assembly descriptors for all of them\n",
    "assemblies_and_messages = genome_api.assembly_descriptors_by_taxon(\n",
    "    taxon=taxid,\n",
    "    limit='all',\n",
    "    filters_refseq_only=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And then we can organize the assemblies by the year they were annotated, and plot the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsUAAAF0CAYAAADLtURVAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/d3fzzAAAACXBIWXMAAAsTAAALEwEAmpwYAAAoF0lEQVR4nO3df3DU9YH/8dfuxiQkIYFNIDZBxYipNvKrBMViAWE7nUHl0ladO9ErClLkhAE8FauCPUUiGLCgYgsWO9XrDz2NlGOOm5ACowxn0iQj4PDr8CcUQrJJSAiJyebz/YNjv6ZEEzabfN7h/XzMOGOWzea5y8K+snyy63EcxxEAAABgMa/bAQAAAIDbGMUAAACwHqMYAAAA1mMUAwAAwHqMYgAAAFiPUQwAAADrxbgdcM6xY8eidllpaWmqqqqK2uVFk8ltktl9tEXO5D7aImdyn8ltktl9tEXO5D6T2ySz+6LdlpGR0eHpPFMMAAAA6zGKAQAAYD1GMQAAAKxnzDHFf89xHDU1NamtrU0ej+eCPvfEiRNqbm7uobLu6azNcRx5vV7Fx8df8PUGAABAZIwdxU1NTbrkkksUE3PhiTExMfL5fD1Q1X1daWttbVVTU5P69evXS1UAAAB2M/bwiba2togG8cUgJiZGbW1tbmcAAABYw9hRbPuhA7ZffwAAgN5k7CgGAAAAekufOT4hdP+0rp+3C+fxrd8UeUyUrF+/XnfffTfHDgMAALiMZ4pdtGHDBp05c8btDAAAAOsxijvx5ptvKhAIKBAIaN68efr88891xx13KBAI6M4779TRo0clSQsWLNDmzZvDn3f11VdLknbt2qXbb79d999/vyZMmKAHHnhAjuPo1Vdf1YkTJ3THHXfo9ttvd+W6AQAA4Kw+c/iEGw4cOKBf/vKX2rRpk/x+v2pqarRgwQLdcccduvPOO/WHP/xBTz75pH7zm9984+Xs3btXxcXFuvTSS5WXl6eSkhLNnDlTv/71r/Xmm2/K7/f30jUCAABAR3im+Bu8//77uvXWW8OjdeDAgfrrX/+qH/3oR5Kkn/zkJ/rggw86vZxRo0YpIyNDXq9X1113nT7//PMe7QYAAMCFYRRHyVdfW7itrU0tLS3hX4uNjQ3/v8/nU2tra6/3AQAA4Otx+MQ3GD9+vGbOnKnZs2eHD5/Izc3Vu+++q9tvv11vv/22brjhBknSkCFDtGfPHk2bNk3//d//3W4Uf52kpCQ1NDRw+AQAAOgRXX31rhNdOI8Jr9zVk/rMKL6Q34iYmJioPBv77W9/W/Pnz9ftt98ePvThmWee0cKFC/XKK6/I7/dr9erVkqTp06fr3nvvVSAQ0M0336yEhIROL3/69OmaPn260tPT9dZbb3W7FwAAAJHxOI7juB0hSceOHWv3cWNjY5eGZUeiNYp7QlfbunP9uyMtLU1VVVW9/nW7grbImdxHW+RM7jO5TTK7j7bImdznVtuFvM9DZ9x6pjjat11GRkaHp3NMMQAAAKzHKAYAAID1jB3FhhzV4Rrbrz8AAEBvMnYUe71eY48L7mmtra3yeo39rQEAALjoGPvqE/Hx8WpqalJzc7M8Hs8FfW5cXJyam5t7qKx7OmtzHEder1fx8fG9WAUAAGA3Y0exx+NRv379IvpcfvoUAAAAF4J/owcAAID1uvRM8b/8y78oPj5eXq9XPp9P+fn5amho0OrVq3Xy5EkNGjRICxcuVFJSkhzH0caNG1VeXq64uDjNnTtXWVlZPX09AAAAgIh1+fCJpUuXKjk5OfxxYWGhhg8frry8PBUWFqqwsFB33323ysvLdfz4ca1Zs0aHDh3Shg0b9Oyzz/ZIPAAAABANER8+UVJSookTJ0qSJk6cqJKSEklSaWmpJkyYII/Ho+zsbJ0+fVo1NTXRqQUAAAB6QJefKV62bJkk6Qc/+IECgYDq6uo0cOBASdKAAQNUV1cnSQoGg0pLSwt/XmpqqoLBYPi85xQVFamoqEiSlJ+f3+5zuismJiaqlxdNJrdJZvfRFjmT+2iLnMl9JrdJZvfRFjmT+9xqOxHFy3Lrtu2t265Lo/jpp5+W3+9XXV2dnnnmmfPeM9rj8Vzwy6YFAgEFAoHwx9F8RQaTX+HB5DbJ7D7aImdyH22RM7nP5DbJ7D7aImdyn8ltXeVWf7Rvu7/fsed06fAJv98vSUpJSdHYsWN1+PBhpaSkhA+LqKmpCR9v7Pf724VXV1eHPx8AAAAwUaejuKmpSWfOnAn//4cffqjLL79cubm52rFjhyRpx44dGjt2rCQpNzdXO3fulOM4OnjwoBISEs47dAIAAAAwSaeHT9TV1en555+XJIVCId10000aNWqUrrrqKq1evVrFxcXhl2STpNGjR6usrEzz589XbGys5s6d27PXAAAAAOimTkdxenq6Vq5ced7p/fv315IlS8473ePxaNasWdGpAwAAAHoB72gHAAAA6zGKAQAAYD1GMQAAAKzHKAYAAID1GMUAAACwHqMYAAAA1mMUAwAAwHqMYgAAAFiPUQwAAADrMYoBAABgPUYxAAAArMcoBgAAgPUYxQAAALAeoxgAAADWYxQDAADAeoxiAAAAWI9RDAAAAOsxigEAAGA9RjEAAACsxygGAACA9RjFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYD1GMQAAAKzHKAYAAID1GMUAAACwHqMYAAAA1mMUAwAAwHqMYgAAAFiPUQwAAADrMYoBAABgPUYxAAAArMcoBgAAgPUYxQAAALAeoxgAAADWYxQDAADAeoxiAAAAWI9RDAAAAOsxigEAAGA9RjEAAACsxygGAACA9RjFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYD1GMQAAAKwX09UztrW1afHixfL7/Vq8eLEqKyv1wgsvqL6+XllZWZo3b55iYmLU0tKiF198UUeOHFH//v21YMECDR48uCevAwAAANAtXX6meMuWLcrMzAx//Prrr+uWW27R2rVrlZiYqOLiYklScXGxEhMTtXbtWt1yyy164403ol8NAAAARFGXRnF1dbXKyso0ZcoUSZLjONq3b5/GjRsnSZo0aZJKSkokSaWlpZo0aZIkady4cdq7d68cx+mBdAAAACA6unT4xGuvvaa7775bZ86ckSTV19crISFBPp9PkuT3+xUMBiVJwWBQqampkiSfz6eEhATV19crOTm53WUWFRWpqKhIkpSfn6+0tLToXCNJMTExUb28aDK5TTK7j7bImdxHW+RM7jO5TTK7j7bImdznVtuJKF6WW7dtb912nY7iv/71r0pJSVFWVpb27dsXtS8cCAQUCATCH1dVVUXtstPS0qJ6edFkcptkdh9tkTO5j7bImdxncptkdh9tkTO5z+S2rnKrP9q3XUZGRoendzqKDxw4oNLSUpWXl+vLL7/UmTNn9Nprr6mxsVGhUEg+n0/BYFB+v1/S2WeNq6urlZqaqlAopMbGRvXv3z9qVwQAAACItk5H8V133aW77rpLkrRv3z79+c9/1vz587Vq1Srt3r1b48eP1/bt25WbmytJGjNmjLZv367s7Gzt3r1bOTk58ng8PXstAADARSt0/7Quna+rhwr41m+KPAYXrYhfp3j69OnavHmz5s2bp4aGBk2ePFmSNHnyZDU0NGjevHnavHmzpk+fHrVYAAAAoCd0+XWKJSknJ0c5OTmSpPT0dC1fvvy888TGxmrRokXRqQMAAAB6Ae9oBwAAAOsxigEAAGA9RjEAAACsxygGAACA9RjFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYD1GMQAAAKzHKAYAAID1GMUAAACwXozbAQAAAH1V6P5pXTrfiS5enm/9pshj0C08UwwAAADrMYoBAABgPUYxAAAArMcoBgAAgPUYxQAAALAeoxgAAADWYxQDAADAeoxiAAAAWI9RDAAAAOsxigEAAGA9RjEAAACsxygGAACA9RjFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYD1GMQAAAKzHKAYAAID1YtwOAAAAgH1C90/r0vlOdPHyfOs3RR4jnikGAAAAGMUAAAAAoxgAAADWYxQDAADAeoxiAAAAWI9RDAAAAOsxigEAAGA9RjEAAACsxygGAACA9RjFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYL2Yzs7w5ZdfaunSpWptbVUoFNK4ceN05513qrKyUi+88ILq6+uVlZWlefPmKSYmRi0tLXrxxRd15MgR9e/fXwsWLNDgwYN747oAAAAAEen0meJLLrlES5cu1cqVK7VixQpVVFTo4MGDev3113XLLbdo7dq1SkxMVHFxsSSpuLhYiYmJWrt2rW655Ra98cYbPX4lAAAAgO7odBR7PB7Fx8dLkkKhkEKhkDwej/bt26dx48ZJkiZNmqSSkhJJUmlpqSZNmiRJGjdunPbu3SvHcXooHwAAAOi+Tg+fkKS2tjY9+uijOn78uH74wx8qPT1dCQkJ8vl8kiS/369gMChJCgaDSk1NlST5fD4lJCSovr5eycnJ7S6zqKhIRUVFkqT8/HylpaVF70rFxET18qLJ5DbJ7D7aImdyH22RM7nP5DbJ7D7aznciypcXzetgcpsU3T6T26Tu93VpFHu9Xq1cuVKnT5/W888/r2PHjnXri0pSIBBQIBAIf1xVVdXtyzwnLS0tqpcXTSa3SWb30RY5k/toi5zJfSa3SWb30dbzTL4OtEWuq30ZGRkdnn5Brz6RmJionJwcHTx4UI2NjQqFQpLOPjvs9/slnX3WuLq6WtLZwy0aGxvVv3//C/kyAAAAQK/qdBSfOnVKp0+flnT2lSg+/PBDZWZmKicnR7t375Ykbd++Xbm5uZKkMWPGaPv27ZKk3bt3KycnRx6Pp4fyAQAAgO7r9PCJmpoavfTSS2pra5PjOLrxxhs1ZswYDRkyRC+88IL+8Ic/6Morr9TkyZMlSZMnT9aLL76oefPmKSkpSQsWLOjp6wAAAAB0S6ej+IorrtCKFSvOOz09PV3Lly8/7/TY2FgtWrQoOnUAAABAL+Ad7QAAAGA9RjEAAACsxygGAACA9RjFAAAAsF6X3rwDAAB0T+j+aV06X1ff5cu3flPkMQDOwzPFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYD1GMQAAAKzHKAYAAID1GMUAAACwHqMYAAAA1mMUAwAAwHqMYgAAAFiPUQwAAADrMYoBAABgPUYxAAAArMcoBgAAgPUYxQAAALAeoxgAAADWYxQDAADAeoxiAAAAWI9RDAAAAOsxigEAAGA9RjEAAACsxygGAACA9RjFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYD1GMQAAAKzHKAYAAID1GMUAAACwHqMYAAAA1mMUAwAAwHqMYgAAAFiPUQwAAADrMYoBAABgPUYxAAAArMcoBgAAgPUYxQAAALAeoxgAAADWYxQDAADAeoxiAAAAWI9RDAAAAOsxigEAAGC9mM7OUFVVpZdeekm1tbXyeDwKBAKaOnWqGhoatHr1ap08eVKDBg3SwoULlZSUJMdxtHHjRpWXlysuLk5z585VVlZWb1wXAAAAICKdPlPs8/l0zz33aPXq1Vq2bJm2bt2qL774QoWFhRo+fLjWrFmj4cOHq7CwUJJUXl6u48ePa82aNZo9e7Y2bNjQ09cBAAAA6JZOR/HAgQPDz/T269dPmZmZCgaDKikp0cSJEyVJEydOVElJiSSptLRUEyZMkMfjUXZ2tk6fPq2ampoevAoAAABA93R6+MRXVVZW6uOPP9awYcNUV1engQMHSpIGDBiguro6SVIwGFRaWlr4c1JTUxUMBsPnPaeoqEhFRUWSpPz8/Haf010xMTFRvbxoMrlNMruPtsiZ3Edb5EzuM7lNcqfvRJQvz43b163fV5NvO5PbpOj2mdwmdb+vy6O4qalJBQUFmjFjhhISEtr9msfjkcfjuaAvHAgEFAgEwh9XVVVd0Od/k7S0tKheXjSZ3CaZ3Udb5Ezuoy1yJveZ3CaZ39cVbvRfDLeb5M5t11W0Ra6rfRkZGR2e3qVXn2htbVVBQYG+//3v64YbbpAkpaSkhA+LqKmpUXJysiTJ7/e3i6qurpbf7+9SJAAAAOCGTkex4zh65ZVXlJmZqVtvvTV8em5urnbs2CFJ2rFjh8aOHRs+fefOnXIcRwcPHlRCQsJ5h04AAAAAJun08IkDBw5o586duvzyy/Xwww9Lkv7pn/5JeXl5Wr16tYqLi8MvySZJo0ePVllZmebPn6/Y2FjNnTu3Z68BAAAA0E2djuJrrrlGf/rTnzr8tSVLlpx3msfj0axZs7pfBgAAAPQS3tEOAAAA1mMUAwAAwHqMYgAAAFiPUQwAAADrMYoBAABgPUYxAAAArMcoBgAAgPU6fZ1iAABw8QvdP63T85zo4mX51m/qXgzgAp4pBgAAgPUYxQAAALAeoxgAAADWYxQDAADAeoxiAAAAWI9RDAAAAOsxigEAAGA9RjEAAACsxygGAACA9RjFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYD1GMQAAAKzHKAYAAID1GMUAAACwHqMYAAAA1mMUAwAAwHqMYgAAAFiPUQwAAADrMYoBAABgPUYxAAAArMcoBgAAgPUYxQAAALAeoxgAAADWYxQDAADAeoxiAAAAWI9RDAAAAOsxigEAAGA9RjEAAACsxygGAACA9RjFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYD1GMQAAAKzHKAYAAID1GMUAAACwXkxnZ3j55ZdVVlamlJQUFRQUSJIaGhq0evVqnTx5UoMGDdLChQuVlJQkx3G0ceNGlZeXKy4uTnPnzlVWVlaPXwkAAACgOzp9pnjSpEn6+c9/3u60wsJCDR8+XGvWrNHw4cNVWFgoSSovL9fx48e1Zs0azZ49Wxs2bOiRaAAAACCaOh3F3/nOd5SUlNTutJKSEk2cOFGSNHHiRJWUlEiSSktLNWHCBHk8HmVnZ+v06dOqqanpgWwAAAAgejo9fKIjdXV1GjhwoCRpwIABqqurkyQFg0GlpaWFz5eamqpgMBg+71cVFRWpqKhIkpSfn9/u87orJiYmqpcXTSa3SWb30RY5k/toi5zJfSa3Se70nYjy5UW7P5p9JrdJ0e0zuU3i9/VCRDSKv8rj8cjj8Vzw5wUCAQUCgfDHVVVV3U0JS0tLi+rlRZPJbZLZfbRFzuQ+2iJncp/JbZL5fV1hcr/JbZLZfbRFrqt9GRkZHZ4e0atPpKSkhA+LqKmpUXJysiTJ7/e3C6qurpbf74/kSwAAAAC9JqJRnJubqx07dkiSduzYobFjx4ZP37lzpxzH0cGDB5WQkNDhoRMAAACASTo9fOKFF17QRx99pPr6es2ZM0d33nmn8vLytHr1ahUXF4dfkk2SRo8erbKyMs2fP1+xsbGaO3duj18BAAAAoLs6HcULFizo8PQlS5acd5rH49GsWbO6HQUAAAD0Jt7RDgAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYD1GMQAAAKzX7bd5BgDYI3T/tC6d70QXL8+3flPkMQAQRTxTDAAAAOsxigEAAGA9RjEAAACsxygGAACA9RjFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYD3e5hkAcNGI5ttQ8xbUgF14phgAAADWYxQDAADAeoxiAAAAWI9RDAAAAOsxigEAAGA9RjEAAACsxygGAACA9RjFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA68W4HQAAbgjdP63T85zo4mX51m/qXgwAwHU8UwwAAADrMYoBAABgPUYxAAAArMcoBgAAgPX4QTsAMExXfghQ4gcBASCaeKYYAAAA1uOZYqAPM/kZRZPbAAD4ezxTDAAAAOsxigEAAGA9RjEAAACsxygGAACA9RjFAAAAsB6jGAAAANZjFAMAAMB6jGIAAABYj1EMAAAA6zGKAQAAYD1GMQAAAKzHKAYAAID1YnriQisqKrRx40a1tbVpypQpysvL64kvgwsQun9al853ogvn8a3f1L0YAAAAw0R9FLe1tenVV1/VE088odTUVD322GPKzc3VkCFDov2lgB4XzW8mJL6hAADAVFEfxYcPH9all16q9PR0SdL3vvc9lZSUMIrxtRieAADAbR7HcZxoXuDu3btVUVGhOXPmSJJ27typQ4cOaebMme3OV1RUpKKiIklSfn5+NBMAAACAC+LaD9oFAgHl5+f3yCBevHhx1C8zWkxuk8zuoy1yJvfRFjmT+0xuk8zuoy1yJveZ3CaZ3ddbbVEfxX6/X9XV1eGPq6ur5ff7o/1lAAAAgKiJ+ii+6qqr9Le//U2VlZVqbW3Vrl27lJubG+0vAwAAAERN1H/Qzufz6b777tOyZcvU1tamm2++WZdddlm0v8w3CgQCvfr1LoTJbZLZfbRFzuQ+2iJncp/JbZLZfbRFzuQ+k9sks/t6qy3qP2gHAAAA9DW8ox0AAACsxygGAACA9RjFAAAAsB6jGAAAANa7KEfxv//7v7udgIvcRx99pGPHjkmS9u/fr02bNqmsrMzlKgAAEKmovyRbb/vNb35z3mk7d+5UU1OTJOm+++7r7aSvVVlZqY8//lhDhgxRZmamqy1VVVVKTk5WbGysHMfR9u3bw21TpkyRz+dzta+0tFQjRoxQbGysqx0dee2113T48GGFQiGNHDlSe/fu1ahRo/Sf//mf2rdvn+655x5X+5qamlRRUaGqqip5vV5lZGRoxIgR8nrN+B746NGjKikpUTAYlHT2DX9yc3M1ZMgQl8u+3l/+8hfdfPPNbmfo6NGjCgaDuvrqqxUfHx8+vaKiQqNGjXIvTNLhw4clScOGDdMXX3yhiooKZWRk6Lvf/a6rXR158cUX9eCDD7qd0aH9+/fr8OHDuuyyyzRy5EhXWw4dOqTMzEwlJCToyy+/VGFhoY4cOaIhQ4boxz/+sRISElzt27Jli66//nqlpaW52tGR1tZWvf/++xo4cKBGjBih9957TwcOHFBmZqYCgYBiYtydXydOnND//M//qLq6Wl6vV9/61rd00003uf57eo5bjxN9/iXZHnjgAV177bUaOXKkzl2V3/3ud+FhMmnSJNfaVqxYoUceeUSSVFJSotdee005OTk6cOCAfvSjH7na9tBDD+nZZ59VXFycXn/9dZ04cUJjx47V3r17JUlz5851rU2Spk+frvj4eI0aNUrjx4/XqFGjjBl1ixYtUkFBgb788kvNmTNHr7zyiuLi4tTa2qpHH31UBQUFrrXt2rVLf/7zn3XFFVdo3759ys7OluM4+uyzzzR//nxdfvnlrrVJUmFhod5//32NHz8+/E6XwWAwfFpeXp6rfV/ngQce0Lp161xt2LJli7Zu3arMzEx9+umnmjFjhsaOHStJevTRR/Xcc8+51vbmm2+qoqJCoVBII0aM0KFDh5STk6M9e/Zo5MiR+vGPf+xa29/fLo7jaN++fbruuusknb3t3PTYY49p+fLlkqSioiJt3bpV119/vT788EONGTPG1T8TixYt0sqVK+Xz+fSrX/1KcXFxGjdunPbs2aNPP/1U//qv/+pamyT99Kc/VXx8vNLT0zV+/HjdeOONSk5OdrXpnDVr1igUCqm5uVmJiYlqamrSDTfcoD179shxHFe/KduyZYvKysp07bXXqry8XEOHDlViYqI++OADzZo1Szk5Oa61Se4+TvT5Z4pXrVqlP/7xj6qoqNA999wjv9+vt956y9XBeU5VVVX4/999910tXbpUgwcP1qlTp/T000+72tjW1qa4uDhJ0p49e7R8+XJ5vV5NmDBBDz/8sGtd52RmZmrJkiXavXu3Nm/erHXr1mns2LG66aab9J3vfMfVNo/HE/7v3MeS5PV61dbW5maa3n77bS1btkxxcXE6deqU1q5dq8cff1yffvqpfv3rX+uZZ55xte8vf/mLCgoKznuW5NZbb9WiRYtcHQBf9wDvOI7q6up6ueZ827Zt03PPPaf4+HhVVlZq1apVOnnypKZOnSq3n9vYvXu3Vq5cqZaWFs2ePVvr1q1TQkKCpk2bpp///OeujuJgMKjMzExNmTJFHo9HjuPoyJEjuu2221xr+qpQKBT+/23btunJJ59UcnKybrvtNj3++OOu/plwHCf8r4ZHjhwJf4NxzTXXGPE4kZ6ervz8fO3Zs0e7du3Sn/70J2VlZWn8+PG64YYb1K9fP9faPvvsMz3//PMKhUKaM2eOfvWrX8nr9er73/++67fdtm3btHLlSnm9Xt16661avny5nnrqKf3gBz/QihUrtGLFClf73Hyc6POjuF+/fpoxY4aOHDmitWvXavTo0a4/QJxzbixJZ//iGzx4sCQpOTm53a+5IS0tTXv37tV1112nQYMGqbq6WoMGDVJ9fb2rXed4PB4lJSUpEAgoEAiotrZWu3bt0htvvKFgMOjqs3ajR4/WkiVL1NLSosmTJ2vVqlXKzs7WRx99pGuvvda1Lunsg9i5Q07i4+PDY+6KK67QmTNn3EyTdPb3taamRoMGDWp3ek1Njet/Jurq6vT4448rMTGx3emO4+jJJ590qap9x7lDJgYPHqynnnpKBQUFOnnypOt/5/l8Pnm9XsXFxSk9PT38T7CxsbGu/74uX75cW7Zs0dtvv6177rlHQ4cOVWxsrOvfXJ/jOI4aGhrkOI4cxwk/0xkfH+/6YWyXXXZZ+NChK664Qv/7v/+rq666SseOHXP9n/+ls3+feL1ejRw5UiNHjlRra6sqKir03nvv6Xe/+51effVV19ocx1Fra6uamprU3NysxsZGJSUlqaWlpd03Qm4JhULyer1qaWkJH26alpZmRJubjxPu36ujJCsrS0uWLNHWrVv17W9/2+0cSdInn3yin/70p3IcRy0tLaqpqdHAgQPV2trq+jOKP/vZz/TSSy/pzTffVL9+/fTII49o6NChOn36tP75n//Z1TZJ5z3IDxgwQFOnTtXUqVN18uRJl6rOuvvuu3Xw4EFJUnZ2to4fP64PPvhAkydP1rhx41xtGz16tJ599llde+21qqioCPece9B124wZM/Rv//Zv+ta3vqXU1FRJZ/9F5fjx45o5c6arbd/97nfV1NSkoUOHnvdrJgyolJQUffLJJ+G++Ph4LV68WOvWrdNnn33maltMTIyam5sVFxen/Pz88OmNjY2uH/Z07tmwG2+8Ub/97W+VkpJixAP/OY2NjVq8eLEcxwmPgYEDB6qpqcn1P7Nz5szRxo0b9fbbb6t///564oknlJqaqtTUVP3sZz9ztU06/3EiJiZGubm5ys3NVXNzs0tVZ918881asGCB2tra9I//+I9atWqVBg8erEOHDul73/ueq21TpkzRY489pmHDhmn//v36h3/4B0nSqVOnlJSU5Gqb5O7jRJ8/pvic2tradgdkDxgwwN2gb3D69GkdPXpU2dnZbqfoiy++0N/+9jeFQiGlpqbqqquucv1BTJL27dvn+nFNnTH1PldWVqYvvvhCQ4cO1YgRIySdPVwmFArpkksucbnubMvhw4fb3XbDhg0z4n5nsurqavl8vg7vZ/v379c111zT+1H/p6WlpcP71qlTp1RbW+v6sexfVVZWpv379+uuu+5yO+UbNTc3q66uLvwvjG5qbGxUZWWl2trajPq77tixY8rIyHA742t99e+406dPa8+ePUpLS9OwYcNcLpM+//xzHT16VJdddpnrP/jfEbceJ/r8KP7kk0+0fv16NTY2hg/Irq6uVmJiombOnKmsrCyXC80dT5LZbZKZfdznekZTU1O7V1Qwicltktl9tEXO5D6T2ySz+2gzV58fxQ8//LBmz56tq6++ut3pBw8e1Pr167Vy5UqXyqSPP/5YGzZs6HA8zZo1S1deeaVrbaYPO5P7TL7PmXy7dcaEV3j4Oia3SWb30RY5k/tMbpPM7qPtm537wfBgMKhRo0Zp+vTp4cM6vvpqLT2hzx9T3NzcfN44kc4e63nu4HG3vPzyy187nl5++WVXx9NLL730tW3r1q1ztU0yu8/k+5zJt5skbd68ucPTHcdx/bYzuU0yu4+2yJncZ3KbZHYfbZHbsGGD7rjjDl199dXatm2blixZokceeUSXXnppj/88QJ8fxaNGjdLy5cs1ceLE8AHZ1dXV2rFjh+svZm/yeDK5TTK7j/tc5H7/+9/rtttu6/Cn6t3+RyuT2ySz+2iLnMl9JrdJZvfRFrmmpqbwY+m0adOUlZWlZ599Vg8++CCvPtGZ++67T+Xl5ee988kPf/hD199JyeTxZHKbZHYf97nIXXnllbr++us7PIyjuLjYhaL/z+Q2yew+2iJncp/JbZLZfbR1T2NjY/ilHa+77jo99NBDKigoUENDQ49+3T5/TLHpOhpPubm5ro8nyew2yfw+U5l8ux07dkxJSUkdvutUbW2tqz8QaHKbZHYfbZEzuc/kNsnsPtoi995772nw4MHnvUJXVVWV3nrrLc2ZM6fHvnafH8WNjY165513VFpaqtraWnk8HqWkpCg3N1d5eXnnvRA/0F3c5wAAuPj0+VG8bNky5eTkaNKkSeHvbmpra7V9+3bt3btXTzzxhGttJo8nk9tM7+M+1/2+kpIS1dXVGdVncpvpfbRdnH0mt5neR1sf7XP6uPnz50f0a73hmWeecd555x2npqYmfFpNTY3zzjvvOE8//bR7YY7ZbY5jdh/3uciZ3Gdym+OY3Udb5EzuM7nNcczuoy1ybvb1+beQGjRokN59913V1taGT6utrVVhYaHS0tLcC5NUWVmpvLy8dsfnDBgwQHl5ea6/VbHJbZLZfdznImdyn8ltktl9tEXO5D6T2ySz+2iLnJt9fX4UL1iwQPX19Xrqqad077336t5779UvfvELNTQ0aOHCha62mTyeTG6TzO7jPhc5k/tMbpPM7qMtcib3mdwmmd1HW+Tc7OvzxxRL0tGjR1VdXa3s7Ox2b09YUVHh6stQNTQ0qLCwUKWlpaqrq5N09rudMWPGKC8vL/wOLbT1vT7ucxdfn8ltpvfRdnH2mdxmeh9tfbOvz4/iLVu2aOvWrcrMzNSnn36qGTNmaOzYsZKkRx99VM8995yrfaaOJ8nsNsncPu5z3WNyn8ltktl9tEXO5D6T2ySz+2iLnFt9ff7wiW3btum5557TI488oqVLl+o//uM/tGXLFknuvzPLli1btGLFCv3Xf/2XHnroIZWUlIR/7fe//72LZWa3SWb3cZ+LnMl9JrdJZvfRFjmT+0xuk8zuoy1ybvb1+Xe0cxwn/F3E4MGD9dRTT6mgoEAnT550faCcG0/x8fGqrKzUqlWrdPLkSU2dOpW2PtzHfe7i7DO5zfQ+2i7OPpPbTO+jrW/29flnilNSUvTJJ5+EP46Pj9fixYtVX1+vzz77zL0wdTyeysvL9dvf/tb1O57JbZLZfdznImdyn8ltktl9tEXO5D6T2ySz+2iLnJt9fX4UP/jgg+e9JaHP59ODDz6oX/ziF+5E/R+Tx5PJbZLZfdznImdyn8ltktl9tEXO5D6T2ySz+2iLnJt9ff4H7UxWXV0tn8/X4fuI79+/X9dcc03vR/0fk9sk8/tMZfrtZnKfyW2S2X20Rc7kPpPbJLP7aIucm32MYgAAAFivzx8+AQAAAHQXoxgAAADWYxQDAADAeoxiAAAAWO//AYw9iEICWjD3AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 864x432 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def annotation_year(annotation_metadata):\n",
    "    return datetime.strptime(annotation_metadata.release_date, '%b %d, %Y').year\n",
    "\n",
    "annots_by_year = Counter()\n",
    "no_annot_assms = []\n",
    "for d in map(lambda d: d.assembly, assemblies_and_messages.assemblies):\n",
    "    if not d.annotation_metadata:\n",
    "        no_annot_assms.append(d.assembly_accession)\n",
    "        continue\n",
    "    annots_by_year[annotation_year(d.annotation_metadata)] += 1\n",
    "\n",
    "\n",
    "if len(no_annot_assms) > 0:\n",
    "    print(dedent(f'''\n",
    "    WARNING!\n",
    "    Some assemblies do not have annotation. \n",
    "    Most likely, this is because of an indexing delay. Skipping {len(no_annot_assms)} assemblies.\n",
    "'''))\n",
    "\n",
    "df = pd.DataFrame.from_dict(annots_by_year, orient='index', columns=['count']).sort_index()\n",
    "df.plot(kind='bar', y='count', figsize=(12,6))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or, we can select just the assemblies that were annotated in 2020:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "332 assemblies were annotated in 2020.\n"
     ]
    }
   ],
   "source": [
    "year_of_interest = 2020\n",
    "assm_accs = []\n",
    "for d in map(lambda d: d.assembly, assemblies_and_messages.assemblies):\n",
    "    if d.annotation_metadata:\n",
    "        if annotation_year(d.annotation_metadata) == year_of_interest:\n",
    "            assm_accs.append(d.assembly_accession)\n",
    "        \n",
    "print(f'{len(assm_accs)} assemblies were annotated in {year_of_interest}.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We could download all of the returned assemblies, based on the accessions stored in `assm_accs`.  If there are quite a few of them, a `download_api.download_assembly_package_post()` request would be required (see docs here)[https://github.com/ncbi/datasets/blob/master/client_docs/python/docs/DownloadApi.md#download_assembly_package_post]. \n",
    "\n",
    "For demonstration purposes, we'll just download the first five assemblies through a GET request."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Download first 5 accessions: ['GCF_012848655.1', 'GCF_013867605.1', 'GCF_013867555.1', 'GCF_013342945.1', 'GCF_012843665.1']\n",
      "Download complete: ncbi_genomes.zip\n",
      "Archive:  ncbi_genomes.zip\n",
      "  Length      Date    Time    Name\n",
      "---------  ---------- -----   ----\n",
      "      661  09-02-2020 13:09   README.md\n",
      "  2243496  09-02-2020 13:09   ncbi_dataset/data/GCF_012843665.1/GCF_012843665.1_ASM1284366v1_genomic.fna\n",
      "  1213074  09-02-2020 13:09   ncbi_dataset/data/GCF_012843665.1/genomic.gff\n",
      "   796745  09-02-2020 13:09   ncbi_dataset/data/GCF_012843665.1/protein.faa\n",
      "  1705083  09-02-2020 13:09   ncbi_dataset/data/GCF_012848655.1/GCF_012848655.1_ASM1284865v1_genomic.fna\n",
      "   861530  09-02-2020 13:09   ncbi_dataset/data/GCF_012848655.1/genomic.gff\n",
      "   589033  09-02-2020 13:09   ncbi_dataset/data/GCF_012848655.1/protein.faa\n",
      "  2016963  09-02-2020 13:09   ncbi_dataset/data/GCF_013342945.1/GCF_013342945.1_ASM1334294v1_genomic.fna\n",
      "  1008359  09-02-2020 13:09   ncbi_dataset/data/GCF_013342945.1/genomic.gff\n",
      "   715800  09-02-2020 13:09   ncbi_dataset/data/GCF_013342945.1/protein.faa\n",
      "  1979975  09-02-2020 13:09   ncbi_dataset/data/GCF_013867555.1/GCF_013867555.1_ASM1386755v1_genomic.fna\n",
      "  1017405  09-02-2020 13:09   ncbi_dataset/data/GCF_013867555.1/genomic.gff\n",
      "   687578  09-02-2020 13:09   ncbi_dataset/data/GCF_013867555.1/protein.faa\n",
      "  1992747  09-02-2020 13:09   ncbi_dataset/data/GCF_013867605.1/GCF_013867605.1_ASM1386760v1_genomic.fna\n",
      "  1028401  09-02-2020 13:09   ncbi_dataset/data/GCF_013867605.1/genomic.gff\n",
      "   711901  09-02-2020 13:09   ncbi_dataset/data/GCF_013867605.1/protein.faa\n",
      "     1735  09-02-2020 13:09   ncbi_dataset/data/GCF_012843665.1/data_report.yaml\n",
      "     1653  09-02-2020 13:09   ncbi_dataset/data/GCF_013342945.1/data_report.yaml\n",
      "     1775  09-02-2020 13:09   ncbi_dataset/data/GCF_013867605.1/data_report.yaml\n",
      "     1659  09-02-2020 13:09   ncbi_dataset/data/GCF_012848655.1/data_report.yaml\n",
      "     1777  09-02-2020 13:09   ncbi_dataset/data/GCF_013867555.1/data_report.yaml\n",
      "    24762  09-02-2020 13:09   ncbi_dataset/data/GCF_012843665.1/sequence_report.yaml\n",
      "     1125  09-02-2020 13:09   ncbi_dataset/data/GCF_013342945.1/sequence_report.yaml\n",
      "     5596  09-02-2020 13:09   ncbi_dataset/data/GCF_013867605.1/sequence_report.yaml\n",
      "     1130  09-02-2020 13:09   ncbi_dataset/data/GCF_012848655.1/sequence_report.yaml\n",
      "     5384  09-02-2020 13:09   ncbi_dataset/data/GCF_013867555.1/sequence_report.yaml\n",
      "     2914  09-02-2020 13:09   ncbi_dataset/data/dataset_catalog.json\n",
      "        0  09-02-2020 13:09   ncbi_dataset/fetch.txt\n",
      "---------                     -------\n",
      " 18618261                     28 files\n",
      "CPU times: user 71.3 ms, sys: 38.6 ms, total: 110 ms\n",
      "Wall time: 2.53 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "## download data\n",
    "print(f'Download first 5 accessions: {assm_accs[0:5]}')\n",
    "dl_response = genome_api.download_assembly_package(\n",
    "    assm_accs[0:5],\n",
    "    include_sequence=True,\n",
    "    include_annotation_type=['GENOME_GFF', 'PROT_FASTA'],\n",
    "    _preload_content=False )\n",
    "\n",
    "## write to a zip file \n",
    "zipfn = 'ncbi_genomes.zip'\n",
    "with open(zipfn, 'wb') as f:\n",
    "    f.write(dl_response.data)\n",
    "print(f'Download complete: {zipfn}')\n",
    "!unzip -l {zipfn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Catalog found with metadata for 5 assemblies\n",
      "Assemblies:\n",
      "GCF_012843665.1, GCF_012848655.1, GCF_013342945.1, GCF_013867555.1, GCF_013867605.1\n",
      "Files for type: GFF3\n",
      "defaultdict(<class 'list'>,\n",
      "            {   'GCF_012843665.1': [   'ncbi_dataset/data/GCF_012843665.1/genomic.gff'],\n",
      "                'GCF_012848655.1': [   'ncbi_dataset/data/GCF_012848655.1/genomic.gff'],\n",
      "                'GCF_013342945.1': [   'ncbi_dataset/data/GCF_013342945.1/genomic.gff'],\n",
      "                'GCF_013867555.1': [   'ncbi_dataset/data/GCF_013867555.1/genomic.gff'],\n",
      "                'GCF_013867605.1': [   'ncbi_dataset/data/GCF_013867605.1/genomic.gff']})\n",
      "Files for type: FASTA\n",
      "defaultdict(<class 'list'>, {})\n",
      "Files for type: GENOMIC_NUCLEOTIDE_FASTA\n",
      "defaultdict(<class 'list'>,\n",
      "            {   'GCF_012843665.1': [   'ncbi_dataset/data/GCF_012843665.1/GCF_012843665.1_ASM1284366v1_genomic.fna'],\n",
      "                'GCF_012848655.1': [   'ncbi_dataset/data/GCF_012848655.1/GCF_012848655.1_ASM1284865v1_genomic.fna'],\n",
      "                'GCF_013342945.1': [   'ncbi_dataset/data/GCF_013342945.1/GCF_013342945.1_ASM1334294v1_genomic.fna'],\n",
      "                'GCF_013867555.1': [   'ncbi_dataset/data/GCF_013867555.1/GCF_013867555.1_ASM1386755v1_genomic.fna'],\n",
      "                'GCF_013867605.1': [   'ncbi_dataset/data/GCF_013867605.1/GCF_013867605.1_ASM1386760v1_genomic.fna']})\n",
      "Files for type: PROTEIN_FASTA\n",
      "defaultdict(<class 'list'>,\n",
      "            {   'GCF_012843665.1': [   'ncbi_dataset/data/GCF_012843665.1/protein.faa'],\n",
      "                'GCF_012848655.1': [   'ncbi_dataset/data/GCF_012848655.1/protein.faa'],\n",
      "                'GCF_013342945.1': [   'ncbi_dataset/data/GCF_013342945.1/protein.faa'],\n",
      "                'GCF_013867555.1': [   'ncbi_dataset/data/GCF_013867555.1/protein.faa'],\n",
      "                'GCF_013867605.1': [   'ncbi_dataset/data/GCF_013867605.1/protein.faa']})\n"
     ]
    }
   ],
   "source": [
    "## function to make list of files \n",
    "from ncbi.datasets.v1alpha1 import catalog_pb2\n",
    "import pprint\n",
    "pp = pprint.PrettyPrinter(indent=4)\n",
    "\n",
    "def retrieve_data_catalog(zip_file):\n",
    "    with zipfile.ZipFile(zip_file, 'r') as zip:\n",
    "        data_catalog = json.loads(zip.read('ncbi_dataset/data/dataset_catalog.json'))\n",
    "    print(f\"Catalog found with metadata for {len(data_catalog['assemblies'])} assemblies\")\n",
    "    return data_catalog\n",
    "\n",
    "def get_assemblies(data_catalog):\n",
    "    return [x['accession'] for x in data_catalog['assemblies']]\n",
    "\n",
    "# Temporary hack to support GENOMIC_NUCLEOTIDE_FASTA & PROTEIN_FASTA \n",
    "# which will be present in the next release\n",
    "def get_file_list(data_catalog, desired_filetype):\n",
    "    desired_filetype = desired_filetype.upper()\n",
    "    if desired_filetype not in catalog_pb2.File.FileType.keys():\n",
    "        raise Exception(f'Filetype {desired_filetype} is invalid.')\n",
    "    \n",
    "    files = defaultdict(list)\n",
    "    for asm in data_catalog['assemblies']:\n",
    "        acc = asm['accession']\n",
    "        for f in asm['files']:\n",
    "            filepath = os.path.join('ncbi_dataset', 'data', f['filePath'])\n",
    "            if f['fileType'] == 'FASTA' and desired_filetype in ('GENOMIC_NUCLEOTIDE_FASTA') and filepath.endswith('fna'):\n",
    "                files[acc].append(filepath)\n",
    "                continue\n",
    "            if f['fileType'] == 'GFF3':\n",
    "                if desired_filetype in ('PROTEIN_FASTA') and filepath.endswith('faa'):\n",
    "                    files[acc].append(filepath)\n",
    "                if desired_filetype in ('GFF3') and filepath.endswith('gff'):\n",
    "                    files[acc].append(filepath)\n",
    "                continue\n",
    "            if f['fileType'] == desired_filetype:\n",
    "                files[acc].append(filepath)\n",
    "        \n",
    "    return files\n",
    "\n",
    "data_catalog = retrieve_data_catalog(zipfn)\n",
    "print(f'Assemblies:')\n",
    "print(', '.join(get_assemblies(data_catalog)))\n",
    "for file_type in ['GFF3', 'FASTA', 'GENOMIC_NUCLEOTIDE_FASTA', 'PROTEIN_FASTA']:\n",
    "    file_list = get_file_list(data_catalog, file_type)\n",
    "    print(f'Files for type: {file_type}')\n",
    "    pp.pprint(file_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## setting up files and directories\n",
    "\n",
    "## temporary files; will be deleted at the end\n",
    "temp_dir = 'tempfiles'\n",
    "temp_gff = temp_dir + '/temp.gff'\n",
    "temp_fa = temp_dir + '/temp.fa'\n",
    "\n",
    "!mkdir -p {temp_dir}\n",
    "\n",
    "## final output files \n",
    "genes_fn = 'crispr_genes.fna'\n",
    "prots_fn = 'crispr_proteins.faa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_gff3_db(files_by_assembly, temp_file, zfh):\n",
    "    '''\n",
    "    create gff3 db in memory, per assembly\n",
    "    okay for bacterial assemblies but use caution\n",
    "    when parsing large assemblies like human\n",
    "    '''\n",
    "\n",
    "    db = {}\n",
    "    for assembly_accession, files in files_by_assembly.items():\n",
    "        with open(temp_file, 'wb') as f:\n",
    "            for file in files:\n",
    "                print(f'\\tWrite {file} to {temp_file}')\n",
    "                f.write(zfh.read(file))\n",
    "            db[assembly_accession] = gffutils.create_db(\n",
    "                temp_file,\n",
    "                dbfn = ':memory:',\n",
    "                force=True,\n",
    "                keep_order=True,\n",
    "                merge_strategy='merge',\n",
    "                sort_attribute_values=True)\n",
    "\n",
    "\n",
    "    return db\n",
    "\n",
    "\n",
    "def combine_fasta(files_by_assembly, temp_file, zfh):\n",
    "    '''\n",
    "    Combine fasta for *all* FASTA files, nt & protein\n",
    "    '''\n",
    "    with open(temp_file, 'wb') as f:\n",
    "        for assembly, files in files_by_assembly.items():\n",
    "            for file in files:\n",
    "                print(f'\\tAppending {file} to {temp_file}')\n",
    "                f.write(zfh.read(file))\n",
    "\n",
    "    print(f'Create FASTA object for {temp_file}.')\n",
    "    return Fasta(temp_file, read_long_names=False, duplicate_action='first')\n",
    "    \n",
    "\n",
    "def extract_genes(gff3_db, assemblies, desired_genes):\n",
    "    crispr_order = defaultdict(list)\n",
    "    crispr_genes = {}\n",
    "\n",
    "    for assembly_accession in assemblies:\n",
    "        for gene in gff3_db[assembly_accession].features_of_type('gene'):\n",
    "            gene_name = gene.attributes.get('Name', None)[0]\n",
    "            if gene_name[:4] not in desired_genes:\n",
    "                continue\n",
    "            gene_range = (gene.start, gene.end)\n",
    "            prot_acc = None\n",
    "            if gene.attributes['gene_biotype'][0] == 'protein_coding':\n",
    "                cds = list(gff3_db[assembly_accession].children(gene, featuretype='CDS'))\n",
    "                prot_acc = cds[0].attributes.get('protein_id', None)[0]\n",
    "\n",
    "            crispr_genes[gene_name] = ([gene.chrom, gene.strand, gene_range, prot_acc])\n",
    "            crispr_order[assembly_accession].append(gene_name)\n",
    "    return crispr_genes, crispr_order\n",
    "\n",
    "def write_fasta(fh, defline, content):\n",
    "    fh.write('>' + defline + '\\n')\n",
    "    fh.write(content + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Create GFF3 Database\n",
      "\tWrite ncbi_dataset/data/GCF_012843665.1/genomic.gff to tempfiles/temp.gff\n",
      "\tWrite ncbi_dataset/data/GCF_012848655.1/genomic.gff to tempfiles/temp.gff\n",
      "\tWrite ncbi_dataset/data/GCF_013342945.1/genomic.gff to tempfiles/temp.gff\n",
      "\tWrite ncbi_dataset/data/GCF_013867555.1/genomic.gff to tempfiles/temp.gff\n",
      "\tWrite ncbi_dataset/data/GCF_013867605.1/genomic.gff to tempfiles/temp.gff\n",
      "Extract crispr genes (12 genes)\n",
      "Write genes to crispr_genes.fna\n",
      "Create genomes object\n",
      "\tAppending ncbi_dataset/data/GCF_012843665.1/GCF_012843665.1_ASM1284366v1_genomic.fna to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_012848655.1/GCF_012848655.1_ASM1284865v1_genomic.fna to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_013342945.1/GCF_013342945.1_ASM1334294v1_genomic.fna to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_013867555.1/GCF_013867555.1_ASM1386755v1_genomic.fna to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_013867605.1/GCF_013867605.1_ASM1386760v1_genomic.fna to tempfiles/temp.fa\n",
      "Create FASTA object for tempfiles/temp.fa.\n",
      "Create proteome object\n",
      "\tAppending ncbi_dataset/data/GCF_012843665.1/protein.faa to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_012848655.1/protein.faa to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_013342945.1/protein.faa to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_013867555.1/protein.faa to tempfiles/temp.fa\n",
      "\tAppending ncbi_dataset/data/GCF_013867605.1/protein.faa to tempfiles/temp.fa\n",
      "Create FASTA object for tempfiles/temp.fa.\n",
      "CPU times: user 4.76 s, sys: 39 ms, total: 4.79 s\n",
      "Wall time: 4.87 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "crispr_genes = set(['cas3', 'cse1', 'cse2', 'cas7', 'cas5', 'cas6', 'cas4', 'cas1', 'cas2', 'cas5', 'cas8', 'cas9', 'csn2'])\n",
    "\n",
    "\n",
    "## create empty files to add data to\n",
    "open(genes_fn, 'w').close()\n",
    "open(prots_fn, 'w').close()\n",
    "\n",
    "with zipfile.ZipFile(zipfn, 'r') as zfh:\n",
    "\n",
    "    print('Create GFF3 Database')\n",
    "    gff3_db = create_gff3_db(get_file_list(data_catalog, 'GFF3'), temp_gff, zfh)\n",
    "    \n",
    "    print(f'Extract crispr genes ({len(crispr_genes)} genes)')\n",
    "    crispr_genes, crispr_order = extract_genes(gff3_db, get_assemblies(data_catalog), crispr_genes)\n",
    "    print(f'Write genes to {genes_fn}')\n",
    "\n",
    "    print('Create genomes object')\n",
    "    genomes = combine_fasta(get_file_list(data_catalog, 'GENOMIC_NUCLEOTIDE_FASTA'), temp_fa, zfh)\n",
    "    \n",
    "    # Write all genes to genes_fn\n",
    "    with open(genes_fn, 'w') as fh:\n",
    "        for gene_name, gene_info in crispr_genes.items():\n",
    "            chrom, strand, gene_range, prot_acc = gene_info\n",
    "            reverse_complement = True if strand == '-' else False\n",
    "            gene_fasta = genomes.get_seq(chrom, gene_range[0], gene_range[1], rc=reverse_complement)\n",
    "            write_fasta(fh, f'{gene_fasta.fancy_name}|{gene_name}', gene_fasta.seq)\n",
    "\n",
    "    print('Create proteome object')\n",
    "    proteome = combine_fasta(get_file_list(data_catalog, 'PROTEIN_FASTA'), temp_fa, zfh)\n",
    "    \n",
    "    # Write protein FASTA for each CRISPR gene\n",
    "    with open(prots_fn, 'w') as fh:\n",
    "        for gene_name, gene_info in crispr_genes.items():\n",
    "            chrom, strand, gene_range, prot_acc = gene_info\n",
    "            if prot_acc is not None:\n",
    "                prot_fasta = proteome[prot_acc][:]\n",
    "                write_fasta(fh, f'{prot_fasta.name}|{gene_name}', prot_fasta.seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_012843665.1 ['cas2', 'cas3', 'cas5', 'cas7i', 'cas8a1', 'cas6', 'cas3', 'cas7e', 'cas5e', 'cas6e', 'cas1e', 'cas2e', 'cas6', 'cas10']\n",
      "GCF_012848655.1 ['cas3', 'cas7e', 'cas5e', 'cas6e', 'cas1e', 'cas2e']\n"
     ]
    }
   ],
   "source": [
    "## order of crispr genes in various assemblies\n",
    "for k,v in crispr_order.items():\n",
    "    print(k, v)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
