### Table of Contents

- [Overview of NCBI Datasets command-line interface (CLI)](#overview-of-ncbi-datasets-command-line-interface-cli)
- [Downloading a genome package (default)](#downloading-a-genome-package-default)
	- [accession/bioproject example](#accession)
	- [taxon](#taxon)
- [Customize the data package using flags](#custom)
	- [annotated](#annot)
	- [assembly-source](#source)
	- [assembly-level](#level)
	- [reference](#ref)
- [Large genome download: rehydration](#rehydrate)
- [Parsing the metadata report](#metadata)
- [Downloading gene data packages (including orthologs)](#gene)
- [User question during the call](#question)

	
### Overview of NCBI Datasets command-line interface (CLI)<a id="cli"></a>

While the web interface is very helpful for browsing genomes, there are times when it's more convenient to access genomes through a command-line environment. For example, let's say you are working on your institution's high-performance computing (HPC) system and you need to download dozens (or hundreds of genomes). Even if you're using the Datasets web interface, this would potentially be a two step process: 

1. Download the genome data package locally;
2. Transfer the files to the HPC system.

With the NCBI Datasets command-line interface (CLI), you can do this process in a single step. Our CLI allows users to access not only genomes, but also genes, ortholog sets and virus genomes. 

The program follows a hierarchy that makes it easier for users to select exact which options they would like to use. In addition to the program commands, additional flags are available for filtering the results.

![Datasets CLI schematic](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/datasets_schema_taxonomy.png) 

The NCBI Datasets CLI command structure is very intuitive. If you take a look at the diagram below, you will notice that the commands are built by choosing one option from each vertical rectangle. 

<div style="text-align: center;">
<img src="https://github.com/user-attachments/assets/212db0a8-2e1e-44ec-ac60-d0660064dd22" style="width: 700px;">
</div>

In addition to *datasets*, we also have *dataformat*, a companion tool to explore and convert metadata to TSV or Excel formats. We will cover the *dataformat* command syntax and use in the metadata section.
For the sake of brevity, we will use the command *summary* instead of download for many of the examples. It's very useful to do that when we're interested to see how different parameters would affect the resulting data package.


### Downloading a genome package (default)<a id="genome"></a>	
#### By accession<a id="accession"></a>

- Bioproject (this would download 1100+ genomes from the VGP Bioproject)

```
datasets download genome accession PRJNA489243

```

- Assembly accession

```
datasets download genome accession GCF_011100635.1

```
Let's unzip and look at the files included in this data package.   
All genome data packages have the same structure, so looking at this one will suffice to understand what we would see in a data package with multiple genomes.

```
unzip ncbi_dataset.zip -d GCF_011100635 
tree GCF_011100635

GCF_011100635
├── README.md
├── md5sum.txt
└── ncbi_dataset
    └── data
        ├── GCF_011100635.1
        │   └── GCF_011100635.1_mTriVul1.pri_genomic.fna
        ├── assembly_data_report.jsonl
        └── dataset_catalog.json

4 directories, 5 files

```

#### By taxon (scientific name, common name, NCBI taxid)<a id="taxon"></a>

Let's say that we want to download all frog genomes from NCBI and we try the following command:

```
datasets download genome taxon frogs
```

This would return the following message:

```
Error: The taxonomy name 'frogs' is not exact. Try using one of the suggested taxids:
Hylidae (family, taxid: 8418, tree frogs)
Ranidae (family, taxid: 8397, riparian frogs)
Hyperoliidae (family, taxid: 8412, African reed frogs)
Anura (order, taxid: 8342, frogs and toads)
Phrynobatrachus (genus, taxid: 143461, puddle frogs)
Pseudis (genus, taxid: 43557, swimming frogs)
Pipidae (family, taxid: 8352, tongueless frogs)
Dendrobatidae (family, taxid: 43469, dart-poison frogs)
Leptodactylidae (family, taxid: 8372, southern frogs)
Phyllomedusa (genus, taxid: 8392, leaf frogs)

Use datasets download genome taxon <command> --help for detailed help about a command.
```

In any circunstance where there's any ambiguity regarding the taxonomic name requested, *datasets* will return an error and will recommend users to input a TaxID instead. 

### Customize the data package using flags<a id="custom"></a>

For the following examples, let's use the command *summary* instead of download. We will be able to see the total number of genomes change based on our request without having to download everything. Pay attention to the last part of the JSON output printed to the screen.

- Back to bioproject:

```
datasets summary genome accession PRJNA489243

... "total_count": 1139}

```

- Now let's see what happens when we request only annotated genomes:<a id="annot"></a>

```
datasets summary genome accession PRJNA489243 --annotated

... "total_count": 234}

```

Now let's say we only want RefSeq annotated genomes:<a id="source"></a>

```
datasets summary genome accession PRJNA489243 --annotated --assembly-source refseq

... "total_count": 228}

```

And finally, only genomes at chromosome level<a id="level"></a>

```
datasets summary genome accession PRJNA489243 --annotated --assembly-source refseq --assembly-level chromosome

... "total_count": 223}

```

We could add one more parameter and request only those that are reference genomes<a id="ref"></a>

```
datasets summary genome accession PRJNA489243 --annotated --assembly-source refseq --assembly-level chromosome --reference

... "total_count": 222}
```

### Large genome download: rehydration<a id="rehydrate"></a>

Now that we decided which genomes we want to download, we have two options:  

1. Direct download of all genomes, which is about 120GB.
2. Dehydrated download and rehydration of the data package

A dehydrated package doesn't include any data. It has a file (`fetch.txt`) that holds the location information of the requested data files. To retrieve those files, the option *rehydrate* is invoked on the CLI and the files are retrieved. 

**Advantages of rehydration:**

* It's faster than a regular download 
* It allows users to download the files as gzip during rehydration, thus taking up less storage space;
* It can be resumed (instead of having to restart in case of failure);
* The `fetch.txt` file can be shared, so you can potentially share your data with other collaborators without having to put hundreds of gigabytes of data on a sharing platform;

#### Dehydrated download/rehydration procedure

**Download and unzip:**

```
datasets download genome accession PRJNA489243 --annotated --assembly-source refseq --assembly-level chromosome --include genome,cds --dehydrated --filename vgp_refseq_chrom_annot.zip

unzip vgp_refseq_chrom_annot.zip -d vgp_refseq_chrom_annot
```

**Rehydrate:**

- All files:

```
datasets rehydrate --directory vgp_refseq_chrom_annot
```

- Files that match a certain criteria:

```
datasets rehydrate --directory vgp_refseq_chrom_annot --match cds
```

### Parsing the metadata report<a id="metadata"></a>

The *datasets* CLI has a companion tool called *dataformat* that can be used to convert the JSON/JSON-Lines reports to TSV or Excel.

Let's say that we would like to create a table with assembly accession, scientific name, scaffold N50 and assembly level so we can select which genomes to download. We can use the *summary* command in *datasets* and pipe the output to *dataformat* to create this table:

```
datasets summary genome accession PRJNA489243 --as-json-lines | dataformat tsv genome --fields accession,organism-name,assmstats-scaffold-n50,assminfo-level | column -t -s$'\t'

Assembly Accession  Organism Name                    Assembly Stats Scaffold N50  Assembly Level
GCA_963993115.1     Abramis brama                    42794514                     Chromosome
GCA_963993125.1     Abramis brama                    465966                       Scaffold
GCA_016880875.1     Acanthisitta chloris             1753226                      Scaffold
GCA_016904835.1     Acanthisitta chloris             40827287                     Chromosome
GCF_904848185.1     Acanthopagrus latus              30717905                     Chromosome
GCA_904848175.1     Acanthopagrus latus              208235                       Scaffold
GCA_904848185.1     Acanthopagrus latus              30717905                     Chromosome
GCF_929443795.1     Accipiter gentilis               35025567                     Chromosome
GCA_929443795.2     Accipiter gentilis               35025567                     Chromosome

```

Another option is to use `jq` to parse the reports. Here's an example: 

```
datasets summary genome taxon vertebrates --assembly-source refseq --assembly-level chromosome,complete --as-json-lines | jq -c 'select(.annotation_info.release_date>"2024-11-01")' | dataformat tsv genome --fields organism-name,accession,assminfo-submitter,assmstats-contig-n50,annotinfo-busco-complete,annotinfo-busco-missing | column -t -s$'\t'

Organism Name                Assembly Accession  Assembly Submitter                       Assembly Stats Contig N50  Annotation BUSCO Complete   Annotation BUSCO Missing 
Dendropsophus ebraccatus     GCF_027789765.1     Vertebrate Genomes Project               8808153                    0.95800376                  0.033898305
Globicephala melas           GCF_963455315.2     WELLCOME SANGER INSTITUTE                3290841                    0.97735286                  0.013498313
Haliaeetus albicilla         GCF_947461875.1     WELLCOME SANGER INSTITUTE                4576000                    0.98836654                  0.0080355005
Melopsittacus undulatus      GCF_012275295.1     G10K                                     16636420                   0.98584795                  0.010074358
Narcine bancroftii           GCF_036971445.1     Vertebrate Genomes Project               5195315                    0.9502087                   0.033094812
Oryctolagus cuniculus        GCF_964237555.1     WELLCOME SANGER INSTITUTE                2030165                    0.9767358                   0.015002174
Phaenicophaeus curvirostris  GCF_032191515.1     Brood Parasitic Bird Genomes Consortium  17162444                   0.9804509                   0.015831135

```

### Downloading gene data packages<a id="gene"></a>

In addition to genome data packages, users can also retrieve gene data using the *datasets* cli. You can search for your gene of interest by accession, symbol or gene-id. In the case of symbol, you currently need to specify the taxon of interest. Also, the gene retrieval is at species level.

```
datasets download gene symbol brca1 --taxon dog

unzip ncbi_dataset.zip -d dog-brca1
Archive:  ncbi_dataset.zip
  inflating: dog-brca1/README.md     
  inflating: dog-brca1/ncbi_dataset/data/rna.fna  
  inflating: dog-brca1/ncbi_dataset/data/protein.faa  
  inflating: dog-brca1/ncbi_dataset/data/data_report.jsonl  
  inflating: dog-brca1/ncbi_dataset/data/dataset_catalog.json  
  inflating: dog-brca1/md5sum.txt   
  
```

If you are interested in orthologs annotated in many species, you can use the ortholog option. Using the same example, you would add the `--ortholog` flag and specify the taxonomy range of interest.

```
datasets download gene symbol brca1 --taxon dog --ortholog vertebrates --filename brca1-vertebrates.zip
Collecting 1 gene record [================================================] 100% 1/1
Collecting 441 gene records [================================================] 100% 441/441

unzip brca1-vertebrates.zip -d brca1-vertebrates
Archive:  brca1-vertebrates.zip
  inflating: brca1-vertebrates/README.md  
  inflating: brca1-vertebrates/ncbi_dataset/data/rna.fna  
  inflating: brca1-vertebrates/ncbi_dataset/data/protein.faa  
  inflating: brca1-vertebrates/ncbi_dataset/data/data_report.jsonl  
  inflating: brca1-vertebrates/ncbi_dataset/data/dataset_catalog.json  
  inflating: brca1-vertebrates/md5sum.txt  
  
```

### User questions during the call:<a id="question"></a>

> Can I combine accession and taxonomy searches in any way? For example if I want information on all mammalian genomes in the VGP BioProject. I tried the —search argument, but got a bit stuck


**Best answer:**

`$ datasets summary genome taxon mammalia --as-json-lines | grep 'PRJNA489243' | dataformat tsv genome --fields 'accession' --elide-header | datasets download genome accession --inputfile /dev/stdin`
 
